// @dxos/client/dist/lib/node-esm/echo/index.mjs
import { createRequire } from 'node:module';const require = createRequire(import.meta.url);
import {
  DocAccessor,
  EchoObjectSchema,
  Expando,
  FQ_ID_LENGTH,
  Filter,
  IndexKind,
  OBJECT_ID_LENGTH,
  ObjectVersion,
  Query,
  QueryOptions,
  Queue,
  ReactiveObjectSchema,
  Ref,
  RefArray,
  ResultFormat,
  SPACE_ID_LENGTH,
  SpaceId,
  SpaceMember,
  SpaceMember2,
  SpaceSchema,
  SpaceState,
  TextKind,
  compareForeignKeys,
  createDocAccessor,
  createEmptyEdgeSyncState,
  createObject,
  createQueueDXN,
  createSubscription,
  decodeReference,
  defineObjectMigration,
  fromCursor,
  fullyQualifiedId,
  getMeta,
  getObjectCore,
  getRangeFromCursor,
  getSchema,
  getSource,
  getSpace,
  getSyncSummary,
  getTarget,
  getTextInRange,
  getType,
  getTypename,
  getVersion,
  importSpace,
  isEchoObject,
  isLiveObject,
  isRelation,
  isSpace,
  live,
  loadObjectReferences,
  makeRef,
  parseFullyQualifiedId,
  parseId,
  refFromDXN,
  toCursor,
  toCursorRange,
  updateText
} from "../chunk-HD6NUSQ4.mjs";
import "../chunk-YSDZCRRN.mjs";
import "../chunk-RT5VBVJV.mjs";
import "../chunk-YV6DEKSE.mjs";
export {
  DocAccessor,
  EchoObjectSchema,
  Expando,
  FQ_ID_LENGTH,
  Filter,
  SpaceMember2 as HaloSpaceMember,
  IndexKind,
  OBJECT_ID_LENGTH,
  ObjectVersion,
  Query,
  QueryOptions,
  Queue,
  ReactiveObjectSchema,
  Ref,
  RefArray,
  ResultFormat,
  SPACE_ID_LENGTH,
  SpaceId,
  SpaceMember,
  SpaceSchema,
  SpaceState,
  TextKind,
  compareForeignKeys,
  createDocAccessor,
  createEmptyEdgeSyncState,
  createObject,
  createQueueDXN,
  createSubscription,
  defineObjectMigration,
  fromCursor,
  fullyQualifiedId,
  getMeta,
  getObjectCore,
  getRangeFromCursor,
  getSchema,
  getSource,
  getSpace,
  getSyncSummary,
  getTarget,
  getTextInRange,
  getType,
  getTypename,
  getVersion,
  importSpace,
  decodeReference as internalDecodeReference,
  isEchoObject,
  isLiveObject,
  isRelation,
  isSpace,
  live,
  loadObjectReferences,
  makeRef,
  parseFullyQualifiedId,
  parseId,
  refFromDXN,
  toCursor,
  toCursorRange,
  updateText
};
