// @dxos/protocols/dist/cjs/src/errors/encoding.js
"use strict";
//
// Copyright 2023 DXOS.org
//
Object.defineProperty(exports, "__esModule", { value: true });
exports.decodeError = exports.encodeError = void 0;
const helpers_js_1 = require("./helpers.js");
const encodeError = (err) => {
    if (typeof err === 'object' && (err === null || err === void 0 ? void 0 : err.message)) {
        return {
            name: err.name,
            message: err.message,
            context: err.context,
            stack: err.stack,
        };
    }
    else if (typeof err === 'string') {
        return {
            message: err,
        };
    }
    else {
        return {
            message: JSON.stringify(err),
        };
    }
};
exports.encodeError = encodeError;
const decodeError = (err, { appendStack } = {}) => {
    var _a;
    const error = (0, helpers_js_1.reconstructError)(err);
    if (appendStack) {
        error.stack = ((_a = err.stack) !== null && _a !== void 0 ? _a : `${error.message}\n`) + appendStack;
    }
    else {
        error.stack = err.stack;
    }
    return error;
};
exports.decodeError = decodeError;
