// @dxos/protocols/src/proto/gen/dxos/tracing.ts
import type { Stream, RequestOptions } from "@dxos/codec-protobuf";
import substitutions from "../../substitutions.js";
import * as dxos_agent_dashboard from "./agent/dashboard.js";
import * as dxos_agent_echoproxy from "./agent/echoproxy.js";
import * as dxos_agent_epoch from "./agent/epoch.js";
import * as dxos_agent_functions from "./agent/functions.js";
import * as dxos_bot from "./bot.js";
import * as dxos_client_services from "./client/services.js";
import * as dxos_config from "./config.js";
import * as dxos_devtools_host from "./devtools/host.js";
import * as dxos_devtools_swarm from "./devtools/swarm.js";
import * as dxos_echo_blob from "./echo/blob.js";
import * as dxos_echo_feed from "./echo/feed.js";
import * as dxos_echo_filter from "./echo/filter.js";
import * as dxos_echo_indexing from "./echo/indexing.js";
import * as dxos_echo_metadata from "./echo/metadata.js";
import * as dxos_echo_model_document from "./echo/model/document.js";
import * as dxos_echo_model_messenger from "./echo/model/messenger.js";
import * as dxos_echo_model_text from "./echo/model/text.js";
import * as dxos_echo_object from "./echo/object.js";
import * as dxos_echo_query from "./echo/query.js";
import * as dxos_echo_service from "./echo/service.js";
import * as dxos_echo_snapshot from "./echo/snapshot.js";
import * as dxos_echo_timeframe from "./echo/timeframe.js";
import * as dxos_edge_calls from "./edge/calls.js";
import * as dxos_edge_messenger from "./edge/messenger.js";
import * as dxos_edge_signal from "./edge/signal.js";
import * as dxos_error from "./error.js";
import * as dxos_gravity from "./gravity.js";
import * as dxos_halo_credentials from "./halo/credentials.js";
import * as dxos_halo_credentials_auth from "./halo/credentials/auth.js";
import * as dxos_halo_credentials_greet from "./halo/credentials/greet.js";
import * as dxos_halo_credentials_identity from "./halo/credentials/identity.js";
import * as dxos_halo_invitations from "./halo/invitations.js";
import * as dxos_halo_keyring from "./halo/keyring.js";
import * as dxos_halo_keys from "./halo/keys.js";
import * as dxos_halo_signed from "./halo/signed.js";
import * as dxos_iframe from "./iframe.js";
import * as dxos_keys from "./keys.js";
import * as dxos_mesh_bridge from "./mesh/bridge.js";
import * as dxos_mesh_broadcast from "./mesh/broadcast.js";
import * as dxos_mesh_messaging from "./mesh/messaging.js";
import * as dxos_mesh_muxer from "./mesh/muxer.js";
import * as dxos_mesh_presence from "./mesh/presence.js";
import * as dxos_mesh_protocol from "./mesh/protocol.js";
import * as dxos_mesh_replicator from "./mesh/replicator.js";
import * as dxos_mesh_signal from "./mesh/signal.js";
import * as dxos_mesh_swarm from "./mesh/swarm.js";
import * as dxos_mesh_teleport from "./mesh/teleport.js";
import * as dxos_mesh_teleport_auth from "./mesh/teleport/auth.js";
import * as dxos_mesh_teleport_automerge from "./mesh/teleport/automerge.js";
import * as dxos_mesh_teleport_blobsync from "./mesh/teleport/blobsync.js";
import * as dxos_mesh_teleport_control from "./mesh/teleport/control.js";
import * as dxos_mesh_teleport_gossip from "./mesh/teleport/gossip.js";
import * as dxos_mesh_teleport_notarization from "./mesh/teleport/notarization.js";
import * as dxos_mesh_teleport_replicator from "./mesh/teleport/replicator.js";
import * as dxos_registry from "./registry.js";
import * as dxos_rpc from "./rpc.js";
import * as dxos_service_agentmanager from "./service/agentmanager.js";
import * as dxos_service_publisher from "./service/publisher.js";
import * as dxos_service_supervisor from "./service/supervisor.js";
import * as dxos_service_tunnel from "./service/tunnel.js";
import * as dxos_type from "./type.js";
import * as dxos_value from "./value.js";
import * as example_testing_data from "../example/testing/data.js";
import * as example_testing_rpc from "../example/testing/rpc.js";
import * as google_protobuf from "../google/protobuf.js";
/**
 * Defined in:
 *   {@link file://./../../dxos/tracing.proto}
 */
export enum TraceLevel {
    DISABLED = 0,
    PRODUCTION = 10,
    DEVELOPMENT = 20,
    DEBUG = 30
}
/**
 * Defined in:
 *   {@link file://./../../dxos/tracing.proto}
 */
export interface Resource {
    id: number;
    className: string;
    instanceId: number;
    info: ReturnType<(typeof substitutions)["google.protobuf.Struct"]["decode"]>;
    links?: ResourceLink[];
    metrics?: Metric[];
}
/**
 * Defined in:
 *   {@link file://./../../dxos/tracing.proto}
 */
export interface ResourceLink {
    from: number;
    to: number;
    attributes: ReturnType<(typeof substitutions)["google.protobuf.Struct"]["decode"]>;
}
/**
 * Defined in:
 *   {@link file://./../../dxos/tracing.proto}
 */
export interface Span {
    id: number;
    /**
     * Options:
     *   - proto3_optional = true
     */
    parentId?: number;
    /**
     * Options:
     *   - proto3_optional = true
     */
    resourceId?: number;
    methodName: string;
    startTs: string;
    /**
     * Options:
     *   - proto3_optional = true
     */
    endTs?: string;
    /**
     * Options:
     *   - proto3_optional = true
     */
    error?: dxos_error.Error;
}
/**
 * Defined in:
 *   {@link file://./../../dxos/tracing.proto}
 */
export interface StreamTraceEvent {
    resourceAdded?: StreamTraceEvent.ResourceAdded[];
    resourceRemoved?: StreamTraceEvent.ResourceRemoved[];
    spanAdded?: StreamTraceEvent.SpanAdded[];
    logAdded?: StreamTraceEvent.LogAdded[];
}
export namespace StreamTraceEvent {
    /**
     * Defined in:
     *   {@link file://./../../dxos/tracing.proto}
     */
    export interface ResourceAdded {
        resource: Resource;
    }
    /**
     * Defined in:
     *   {@link file://./../../dxos/tracing.proto}
     */
    export interface ResourceRemoved {
        id: number;
    }
    /**
     * Defined in:
     *   {@link file://./../../dxos/tracing.proto}
     */
    export interface SpanAdded {
        span: Span;
    }
    /**
     * Defined in:
     *   {@link file://./../../dxos/tracing.proto}
     */
    export interface LogAdded {
        log: dxos_client_services.LogEntry;
    }
}
/**
 * Defined in:
 *   {@link file://./../../dxos/tracing.proto}
 */
export interface Metric {
    name: string;
    counter?: Metric.Counter;
    timeSeries?: Metric.TimeSeries;
    multiCounter?: Metric.MultiCounter;
    custom?: Metric.Custom;
}
export namespace Metric {
    /**
     * Defined in:
     *   {@link file://./../../dxos/tracing.proto}
     */
    export interface Counter {
        value: number;
        /**
         * Options:
         *   - proto3_optional = true
         */
        units?: string;
    }
    /**
     * Defined in:
     *   {@link file://./../../dxos/tracing.proto}
     */
    export interface MultiCounter {
        records?: Metric.MultiCounter.Record[];
        /**
         * Options:
         *   - proto3_optional = true
         */
        units?: string;
    }
    export namespace MultiCounter {
        /**
         * Defined in:
         *   {@link file://./../../dxos/tracing.proto}
         */
        export interface Record {
            key: string;
            value: number;
        }
    }
    /**
     * Defined in:
     *   {@link file://./../../dxos/tracing.proto}
     */
    export interface TimeSeries {
        tracks?: Metric.TimeSeries.Track[];
    }
    export namespace TimeSeries {
        /**
         * Defined in:
         *   {@link file://./../../dxos/tracing.proto}
         */
        export interface Point {
            value: number;
        }
        /**
         * Defined in:
         *   {@link file://./../../dxos/tracing.proto}
         */
        export interface Track {
            name: string;
            /**
             * Options:
             *   - proto3_optional = true
             */
            units?: string;
            points?: Metric.TimeSeries.Point[];
            total: number;
        }
    }
    /**
     * Defined in:
     *   {@link file://./../../dxos/tracing.proto}
     */
    export interface Custom {
        payload: ReturnType<(typeof substitutions)["google.protobuf.Struct"]["decode"]>;
    }
}
/**
 * Defined in:
 *   {@link file://./../../dxos/tracing.proto}
 */
export interface TracingService {
    streamTrace: (request: void, options?: RequestOptions) => Stream<StreamTraceEvent>;
}
