// @dxos/protocols/src/proto/gen/google/protobuf.ts
import type { Stream, RequestOptions } from "@dxos/codec-protobuf";
import substitutions from "../../substitutions.js";
import * as dxos_agent_dashboard from "../dxos/agent/dashboard.js";
import * as dxos_agent_echoproxy from "../dxos/agent/echoproxy.js";
import * as dxos_agent_epoch from "../dxos/agent/epoch.js";
import * as dxos_agent_functions from "../dxos/agent/functions.js";
import * as dxos_bot from "../dxos/bot.js";
import * as dxos_client_services from "../dxos/client/services.js";
import * as dxos_config from "../dxos/config.js";
import * as dxos_devtools_host from "../dxos/devtools/host.js";
import * as dxos_devtools_swarm from "../dxos/devtools/swarm.js";
import * as dxos_echo_blob from "../dxos/echo/blob.js";
import * as dxos_echo_feed from "../dxos/echo/feed.js";
import * as dxos_echo_filter from "../dxos/echo/filter.js";
import * as dxos_echo_indexing from "../dxos/echo/indexing.js";
import * as dxos_echo_metadata from "../dxos/echo/metadata.js";
import * as dxos_echo_model_document from "../dxos/echo/model/document.js";
import * as dxos_echo_model_messenger from "../dxos/echo/model/messenger.js";
import * as dxos_echo_model_text from "../dxos/echo/model/text.js";
import * as dxos_echo_object from "../dxos/echo/object.js";
import * as dxos_echo_query from "../dxos/echo/query.js";
import * as dxos_echo_service from "../dxos/echo/service.js";
import * as dxos_echo_snapshot from "../dxos/echo/snapshot.js";
import * as dxos_echo_timeframe from "../dxos/echo/timeframe.js";
import * as dxos_edge_calls from "../dxos/edge/calls.js";
import * as dxos_edge_messenger from "../dxos/edge/messenger.js";
import * as dxos_edge_signal from "../dxos/edge/signal.js";
import * as dxos_error from "../dxos/error.js";
import * as dxos_gravity from "../dxos/gravity.js";
import * as dxos_halo_credentials from "../dxos/halo/credentials.js";
import * as dxos_halo_credentials_auth from "../dxos/halo/credentials/auth.js";
import * as dxos_halo_credentials_greet from "../dxos/halo/credentials/greet.js";
import * as dxos_halo_credentials_identity from "../dxos/halo/credentials/identity.js";
import * as dxos_halo_invitations from "../dxos/halo/invitations.js";
import * as dxos_halo_keyring from "../dxos/halo/keyring.js";
import * as dxos_halo_keys from "../dxos/halo/keys.js";
import * as dxos_halo_signed from "../dxos/halo/signed.js";
import * as dxos_iframe from "../dxos/iframe.js";
import * as dxos_keys from "../dxos/keys.js";
import * as dxos_mesh_bridge from "../dxos/mesh/bridge.js";
import * as dxos_mesh_broadcast from "../dxos/mesh/broadcast.js";
import * as dxos_mesh_messaging from "../dxos/mesh/messaging.js";
import * as dxos_mesh_muxer from "../dxos/mesh/muxer.js";
import * as dxos_mesh_presence from "../dxos/mesh/presence.js";
import * as dxos_mesh_protocol from "../dxos/mesh/protocol.js";
import * as dxos_mesh_replicator from "../dxos/mesh/replicator.js";
import * as dxos_mesh_signal from "../dxos/mesh/signal.js";
import * as dxos_mesh_swarm from "../dxos/mesh/swarm.js";
import * as dxos_mesh_teleport from "../dxos/mesh/teleport.js";
import * as dxos_mesh_teleport_auth from "../dxos/mesh/teleport/auth.js";
import * as dxos_mesh_teleport_automerge from "../dxos/mesh/teleport/automerge.js";
import * as dxos_mesh_teleport_blobsync from "../dxos/mesh/teleport/blobsync.js";
import * as dxos_mesh_teleport_control from "../dxos/mesh/teleport/control.js";
import * as dxos_mesh_teleport_gossip from "../dxos/mesh/teleport/gossip.js";
import * as dxos_mesh_teleport_notarization from "../dxos/mesh/teleport/notarization.js";
import * as dxos_mesh_teleport_replicator from "../dxos/mesh/teleport/replicator.js";
import * as dxos_registry from "../dxos/registry.js";
import * as dxos_rpc from "../dxos/rpc.js";
import * as dxos_service_agentmanager from "../dxos/service/agentmanager.js";
import * as dxos_service_publisher from "../dxos/service/publisher.js";
import * as dxos_service_supervisor from "../dxos/service/supervisor.js";
import * as dxos_service_tunnel from "../dxos/service/tunnel.js";
import * as dxos_tracing from "../dxos/tracing.js";
import * as dxos_type from "../dxos/type.js";
import * as dxos_value from "../dxos/value.js";
import * as example_testing_data from "../example/testing/data.js";
import * as example_testing_rpc from "../example/testing/rpc.js";
export interface Empty {
}
export interface Struct {
    fields?: Partial<Record<string, Value>>;
}
export interface Value {
    nullValue?: NullValue;
    numberValue?: number;
    stringValue?: string;
    boolValue?: boolean;
    structValue?: ReturnType<(typeof substitutions)["google.protobuf.Struct"]["decode"]>;
    listValue?: ListValue;
}
export enum NullValue {
    NULL_VALUE = 0
}
export interface ListValue {
    values?: Value[];
}
export interface Timestamp {
    seconds: string;
    nanos: number;
}
export interface Any {
    type_url: string;
    value: Uint8Array;
}
export interface FileDescriptorSet {
    file?: FileDescriptorProto[];
}
export interface FileDescriptorProto {
    name: string;
    package: string;
    dependency?: string[];
    /**
     * Options:
     *   - packed = false
     */
    publicDependency?: number[];
    /**
     * Options:
     *   - packed = false
     */
    weakDependency?: number[];
    messageType?: DescriptorProto[];
    enumType?: EnumDescriptorProto[];
    service?: ServiceDescriptorProto[];
    extension?: FieldDescriptorProto[];
    options: FileOptions;
    sourceCodeInfo: SourceCodeInfo;
    syntax: string;
}
export interface DescriptorProto {
    name: string;
    field?: FieldDescriptorProto[];
    extension?: FieldDescriptorProto[];
    nestedType?: DescriptorProto[];
    enumType?: EnumDescriptorProto[];
    extensionRange?: DescriptorProto.ExtensionRange[];
    oneofDecl?: OneofDescriptorProto[];
    options: MessageOptions;
    reservedRange?: DescriptorProto.ReservedRange[];
    reservedName?: string[];
}
export namespace DescriptorProto {
    export interface ExtensionRange {
        start: number;
        end: number;
    }
    export interface ReservedRange {
        start: number;
        end: number;
    }
}
export interface FieldDescriptorProto {
    name: string;
    number: number;
    label: FieldDescriptorProto.Label;
    type: FieldDescriptorProto.Type;
    typeName: string;
    extendee: string;
    defaultValue: string;
    oneofIndex: number;
    jsonName: string;
    options: FieldOptions;
}
export namespace FieldDescriptorProto {
    export enum Type {
        TYPE_DOUBLE = 1,
        TYPE_FLOAT = 2,
        TYPE_INT64 = 3,
        TYPE_UINT64 = 4,
        TYPE_INT32 = 5,
        TYPE_FIXED64 = 6,
        TYPE_FIXED32 = 7,
        TYPE_BOOL = 8,
        TYPE_STRING = 9,
        TYPE_GROUP = 10,
        TYPE_MESSAGE = 11,
        TYPE_BYTES = 12,
        TYPE_UINT32 = 13,
        TYPE_ENUM = 14,
        TYPE_SFIXED32 = 15,
        TYPE_SFIXED64 = 16,
        TYPE_SINT32 = 17,
        TYPE_SINT64 = 18
    }
    export enum Label {
        LABEL_OPTIONAL = 1,
        LABEL_REQUIRED = 2,
        LABEL_REPEATED = 3
    }
}
export interface OneofDescriptorProto {
    name: string;
    options: OneofOptions;
}
export interface EnumDescriptorProto {
    name: string;
    value?: EnumValueDescriptorProto[];
    options: EnumOptions;
}
export interface EnumValueDescriptorProto {
    name: string;
    number: number;
    options: EnumValueOptions;
}
export interface ServiceDescriptorProto {
    name: string;
    method?: MethodDescriptorProto[];
    options: ServiceOptions;
}
export interface MethodDescriptorProto {
    name: string;
    inputType: string;
    outputType: string;
    options: MethodOptions;
    clientStreaming: boolean;
    serverStreaming: boolean;
}
export interface FileOptions {
    javaPackage: string;
    javaOuterClassname: string;
    javaMultipleFiles: boolean;
    /**
     * Options:
     *   - deprecated = true
     */
    javaGenerateEqualsAndHash: boolean;
    javaStringCheckUtf8: boolean;
    /**
     * Options:
     *   - default = "SPEED"
     */
    optimizeFor: FileOptions.OptimizeMode;
    goPackage: string;
    ccGenericServices: boolean;
    javaGenericServices: boolean;
    pyGenericServices: boolean;
    deprecated: boolean;
    ccEnableArenas: boolean;
    objcClassPrefix: string;
    csharpNamespace: string;
    uninterpretedOption?: UninterpretedOption[];
}
export namespace FileOptions {
    export enum OptimizeMode {
        SPEED = 1,
        CODE_SIZE = 2,
        LITE_RUNTIME = 3
    }
}
export interface MessageOptions {
    messageSetWireFormat: boolean;
    noStandardDescriptorAccessor: boolean;
    deprecated: boolean;
    mapEntry: boolean;
    uninterpretedOption?: UninterpretedOption[];
}
export interface FieldOptions {
    /**
     * Options:
     *   - default = "STRING"
     */
    ctype: FieldOptions.CType;
    packed: boolean;
    /**
     * Options:
     *   - default = "JS_NORMAL"
     */
    jstype: FieldOptions.JSType;
    lazy: boolean;
    deprecated: boolean;
    weak: boolean;
    uninterpretedOption?: UninterpretedOption[];
    /**
     * Options:
     *   - proto3_optional = true
     */
    ".dxos.config.envVar"?: string;
    /**
     * Options:
     *   - proto3_optional = true
     */
    ".preserveAny"?: boolean;
}
export namespace FieldOptions {
    export enum CType {
        STRING = 0,
        CORD = 1,
        STRING_PIECE = 2
    }
    export enum JSType {
        JS_NORMAL = 0,
        JS_STRING = 1,
        JS_NUMBER = 2
    }
}
export interface OneofOptions {
    uninterpretedOption?: UninterpretedOption[];
}
export interface EnumOptions {
    allowAlias: boolean;
    deprecated: boolean;
    uninterpretedOption?: UninterpretedOption[];
}
export interface EnumValueOptions {
    deprecated: boolean;
    uninterpretedOption?: UninterpretedOption[];
}
export interface ServiceOptions {
    deprecated: boolean;
    uninterpretedOption?: UninterpretedOption[];
}
export interface MethodOptions {
    deprecated: boolean;
    uninterpretedOption?: UninterpretedOption[];
}
export interface UninterpretedOption {
    name?: UninterpretedOption.NamePart[];
    identifierValue: string;
    positiveIntValue: string;
    negativeIntValue: string;
    doubleValue: number;
    stringValue: Uint8Array;
    aggregateValue: string;
}
export namespace UninterpretedOption {
    export interface NamePart {
        namePart: string;
        isExtension: boolean;
    }
}
export interface SourceCodeInfo {
    location?: SourceCodeInfo.Location[];
}
export namespace SourceCodeInfo {
    export interface Location {
        path?: number[];
        span?: number[];
        leadingComments: string;
        trailingComments: string;
        leadingDetachedComments?: string[];
    }
}
export interface GeneratedCodeInfo {
    annotation?: GeneratedCodeInfo.Annotation[];
}
export namespace GeneratedCodeInfo {
    export interface Annotation {
        path?: number[];
        sourceFile: string;
        begin: number;
        end: number;
    }
}
export interface DoubleValue {
    value: number;
}
export interface FloatValue {
    value: number;
}
export interface Int64Value {
    value: string;
}
export interface UInt64Value {
    value: string;
}
export interface Int32Value {
    value: number;
}
export interface UInt32Value {
    value: number;
}
export interface BoolValue {
    value: boolean;
}
export interface StringValue {
    value: string;
}
export interface BytesValue {
    value: Uint8Array;
}
