// @dxos/messaging/dist/lib/node/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var node_exports = {};
__export(node_exports, {
  EdgeSignalManager: () => import_chunk_GMND65DN.EdgeSignalManager,
  MemorySignalManager: () => import_chunk_GMND65DN.MemorySignalManager,
  MemorySignalManagerContext: () => import_chunk_GMND65DN.MemorySignalManagerContext,
  Messenger: () => import_chunk_GMND65DN.Messenger,
  PeerInfoHash: () => import_chunk_GMND65DN.PeerInfoHash,
  SignalClient: () => import_chunk_GMND65DN.SignalClient,
  WebsocketSignalManager: () => import_chunk_GMND65DN.WebsocketSignalManager,
  setIdentityTags: () => import_chunk_GMND65DN.setIdentityTags
});
module.exports = __toCommonJS(node_exports);
var import_chunk_GMND65DN = require("./chunk-GMND65DN.cjs");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  EdgeSignalManager,
  MemorySignalManager,
  MemorySignalManagerContext,
  Messenger,
  PeerInfoHash,
  SignalClient,
  WebsocketSignalManager,
  setIdentityTags
});
