// @dxos/react-client/dist/lib/node/invitations/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var invitations_exports = {};
__export(invitations_exports, {
  AuthenticatingInvitationObservable: () => import_chunk_FYOAVV65.AuthenticatingInvitationObservable,
  CancellableInvitationObservable: () => import_chunk_FYOAVV65.CancellableInvitationObservable,
  Invitation: () => import_chunk_FYOAVV65.Invitation,
  InvitationEncoder: () => import_chunk_FYOAVV65.InvitationEncoder,
  InvitationsProxy: () => import_chunk_FYOAVV65.InvitationsProxy,
  useInvitationStatus: () => import_chunk_FYOAVV65.useInvitationStatus
});
module.exports = __toCommonJS(invitations_exports);
var import_chunk_FYOAVV65 = require("../chunk-FYOAVV65.cjs");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AuthenticatingInvitationObservable,
  CancellableInvitationObservable,
  Invitation,
  InvitationEncoder,
  InvitationsProxy,
  useInvitationStatus
});
