// @dxos/network-manager/dist/lib/browser/index.mjs
import "@dxos/node-std/globals";
import {
  Connection,
  ConnectionLimiter,
  ConnectionLog,
  ConnectionState,
  EventType,
  FullyConnectedTopology,
  MAX_CONCURRENT_INITIATING_CONNECTIONS,
  MMSTTopology,
  MemoryTransport,
  MemoryTransportFactory,
  RtcTransportProxy,
  RtcTransportProxyFactory,
  RtcTransportService,
  StarTopology,
  Swarm,
  SwarmMapper,
  SwarmMessenger,
  SwarmNetworkManager,
  TransportKind,
  createIceProvider,
  createRtcTransportFactory,
  createTeleportProtocolFactory
} from "./chunk-LFKR6BAF.mjs";
export {
  Connection,
  ConnectionLimiter,
  ConnectionLog,
  ConnectionState,
  EventType,
  FullyConnectedTopology,
  MAX_CONCURRENT_INITIATING_CONNECTIONS,
  MMSTTopology,
  MemoryTransport,
  MemoryTransportFactory,
  RtcTransportProxy,
  RtcTransportProxyFactory,
  RtcTransportService,
  StarTopology,
  Swarm,
  SwarmMapper,
  SwarmMessenger,
  SwarmNetworkManager,
  TransportKind,
  createIceProvider,
  createRtcTransportFactory,
  createTeleportProtocolFactory
};
