// @dxos/protocols/dist/cjs/src/proto/gen/dxos/echo/filter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryOptions = void 0;
var QueryOptions;
(function (QueryOptions) {
    /**
     * Defined in:
     *   {@link file://./../../../dxos/echo/filter.proto}
     */
    let ShowDeletedOption;
    (function (ShowDeletedOption) {
        ShowDeletedOption[ShowDeletedOption["HIDE_DELETED"] = 0] = "HIDE_DELETED";
        ShowDeletedOption[ShowDeletedOption["SHOW_DELETED"] = 1] = "SHOW_DELETED";
        ShowDeletedOption[ShowDeletedOption["SHOW_DELETED_ONLY"] = 2] = "SHOW_DELETED_ONLY";
    })(ShowDeletedOption = QueryOptions.ShowDeletedOption || (QueryOptions.ShowDeletedOption = {}));
    /**
     * Defined in:
     *   {@link file://./../../../dxos/echo/filter.proto}
     */
    let DataLocation;
    (function (DataLocation) {
        DataLocation[DataLocation["ALL"] = 0] = "ALL";
        DataLocation[DataLocation["LOCAL"] = 1] = "LOCAL";
        DataLocation[DataLocation["REMOTE"] = 2] = "REMOTE";
    })(DataLocation = QueryOptions.DataLocation || (QueryOptions.DataLocation = {}));
})(QueryOptions || (exports.QueryOptions = QueryOptions = {}));
