// @dxos/protocols/src/buf/proto/gen/dxos/mesh/broadcast_pb.ts
//
// Copyright 2022 DXOS.org
//

// @generated by protoc-gen-es v2.1.0 with parameter "target=ts,import_extension=js"
// @generated from file dxos/mesh/broadcast.proto (package dxos.mesh.broadcast, syntax proto3)
/* eslint-disable */

import type { GenFile, GenMessage } from "@bufbuild/protobuf/codegenv1";
import { fileDesc, messageDesc } from "@bufbuild/protobuf/codegenv1";
import type { Message } from "@bufbuild/protobuf";

/**
 * Describes the file dxos/mesh/broadcast.proto.
 */
export const file_dxos_mesh_broadcast: GenFile = /*@__PURE__*/
  fileDesc("ChlkeG9zL21lc2gvYnJvYWRjYXN0LnByb3RvEhNkeG9zLm1lc2guYnJvYWRjYXN0Ik8KBlBhY2tldBILCgNzZXEYASABKAwSDgoGb3JpZ2luGAIgASgMEhEKBGZyb20YAyABKAxIAIgBARIMCgRkYXRhGAQgASgMQgcKBV9mcm9tYgZwcm90bzM");

/**
 * @generated from message dxos.mesh.broadcast.Packet
 */
export type Packet = Message<"dxos.mesh.broadcast.Packet"> & {
  /**
   * Id message.
   *
   * @generated from field: bytes seq = 1;
   */
  seq: Uint8Array;

  /**
   * Represents the author's ID of the message. To identify a message in the network you should check for the: `seqno + origin`.
   *
   * @generated from field: bytes origin = 2;
   */
  origin: Uint8Array;

  /**
   * Represents the current sender's ID of the message.
   *
   * @generated from field: optional bytes from = 3;
   */
  from?: Uint8Array;

  /**
   * Represents an opaque blob of data, it can contain any data that the publisher wants to send.
   *
   * @generated from field: bytes data = 4;
   */
  data: Uint8Array;
};

/**
 * Describes the message dxos.mesh.broadcast.Packet.
 * Use `create(PacketSchema)` to create a new message.
 */
export const PacketSchema: GenMessage<Packet> = /*@__PURE__*/
  messageDesc(file_dxos_mesh_broadcast, 0);
