// @dxos/protocols/src/buf/proto/gen/dxos/echo/model/messenger_pb.ts
//
// Copyright 2020 DXOS.org
//

// @generated by protoc-gen-es v2.1.0 with parameter "target=ts,import_extension=js"
// @generated from file dxos/echo/model/messenger.proto (package dxos.echo.model.messenger, syntax proto3)
/* eslint-disable */

import type { GenFile, GenMessage } from "@bufbuild/protobuf/codegenv1";
import { fileDesc, messageDesc } from "@bufbuild/protobuf/codegenv1";
import type { Message as Message$1 } from "@bufbuild/protobuf";

/**
 * Describes the file dxos/echo/model/messenger.proto.
 */
export const file_dxos_echo_model_messenger: GenFile = /*@__PURE__*/
  fileDesc("Ch9keG9zL2VjaG8vbW9kZWwvbWVzc2VuZ2VyLnByb3RvEhlkeG9zLmVjaG8ubW9kZWwubWVzc2VuZ2VyIjoKB01lc3NhZ2USEQoJdGltZXN0YW1wGAEgASgJEgwKBHRleHQYAiABKAkSDgoGc2VuZGVyGAMgASgJYgZwcm90bzM");

/**
 * @generated from message dxos.echo.model.messenger.Message
 */
export type Message = Message$1<"dxos.echo.model.messenger.Message"> & {
  /**
   * @generated from field: string timestamp = 1;
   */
  timestamp: string;

  /**
   * @generated from field: string text = 2;
   */
  text: string;

  /**
   * @generated from field: string sender = 3;
   */
  sender: string;
};

/**
 * Describes the message dxos.echo.model.messenger.Message.
 * Use `create(MessageSchema)` to create a new message.
 */
export const MessageSchema: GenMessage<Message> = /*@__PURE__*/
  messageDesc(file_dxos_echo_model_messenger, 0);
