// @dxos/protocols/src/buf/proto/gen/dxos/halo/credentials/identity_pb.ts
//
// Copyright 2020 DXOS.org
//

// @generated by protoc-gen-es v2.1.0 with parameter "target=ts,import_extension=js"
// @generated from file dxos/halo/credentials/identity.proto (package dxos.halo.credentials.identity, syntax proto3)
/* eslint-disable */

import type { GenFile, GenMessage } from "@bufbuild/protobuf/codegenv1";
import { fileDesc, messageDesc } from "@bufbuild/protobuf/codegenv1";
import type { PublicKey } from "../../keys_pb.js";
import { file_dxos_keys } from "../../keys_pb.js";
import type { Message } from "@bufbuild/protobuf";

/**
 * Describes the file dxos/halo/credentials/identity.proto.
 */
export const file_dxos_halo_credentials_identity: GenFile = /*@__PURE__*/
  fileDesc("CiRkeG9zL2hhbG8vY3JlZGVudGlhbHMvaWRlbnRpdHkucHJvdG8SHmR4b3MuaGFsby5jcmVkZW50aWFscy5pZGVudGl0eSJOCgxJZGVudGl0eUluZm8SKAoKcHVibGljX2tleRgBIAEoCzIULmR4b3Mua2V5cy5QdWJsaWNLZXkSFAoMZGlzcGxheV9uYW1lGAIgASgJIkwKCkRldmljZUluZm8SKAoKcHVibGljX2tleRgBIAEoCzIULmR4b3Mua2V5cy5QdWJsaWNLZXkSFAoMZGlzcGxheV9uYW1lGAIgASgJYgZwcm90bzM", [file_dxos_keys]);

/**
 * Additional, descriptive information about an Identity. Must be signed by the Identity's key.
 *
 * @generated from message dxos.halo.credentials.identity.IdentityInfo
 */
export type IdentityInfo = Message<"dxos.halo.credentials.identity.IdentityInfo"> & {
  /**
   * The publicKey of the Identity (must be signed by this key).
   *
   * @generated from field: dxos.keys.PublicKey public_key = 1;
   */
  publicKey?: PublicKey;

  /**
   * The display name to use for this key.
   *
   * @generated from field: string display_name = 2;
   */
  displayName: string;
};

/**
 * Describes the message dxos.halo.credentials.identity.IdentityInfo.
 * Use `create(IdentityInfoSchema)` to create a new message.
 */
export const IdentityInfoSchema: GenMessage<IdentityInfo> = /*@__PURE__*/
  messageDesc(file_dxos_halo_credentials_identity, 0);

/**
 * Additional, descriptive information about a Device. Must be signed by the Device's key.
 *
 * @generated from message dxos.halo.credentials.identity.DeviceInfo
 */
export type DeviceInfo = Message<"dxos.halo.credentials.identity.DeviceInfo"> & {
  /**
   * The publicKey of the Device (must be signed by this key).
   *
   * @generated from field: dxos.keys.PublicKey public_key = 1;
   */
  publicKey?: PublicKey;

  /**
   * The display name to use for this key.
   *
   * @generated from field: string display_name = 2;
   */
  displayName: string;
};

/**
 * Describes the message dxos.halo.credentials.identity.DeviceInfo.
 * Use `create(DeviceInfoSchema)` to create a new message.
 */
export const DeviceInfoSchema: GenMessage<DeviceInfo> = /*@__PURE__*/
  messageDesc(file_dxos_halo_credentials_identity, 1);
