// @dxos/client/dist/lib/node/services-NYQN2ZCW.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var services_NYQN2ZCW_exports = {};
__export(services_NYQN2ZCW_exports, {
  AgentClientServiceProvider: () => import_chunk_GXHU364O.AgentClientServiceProvider,
  AgentManagerClient: () => import_chunk_GXHU364O.AgentManagerClient,
  ClientServicesProxy: () => import_chunk_GXHU364O.ClientServicesProxy,
  FakeAgentHostingProvider: () => import_chunk_GXHU364O.FakeAgentHostingProvider,
  IFrameManager: () => import_chunk_GXHU364O.IFrameManager,
  LocalClientServices: () => import_chunk_GXHU364O.LocalClientServices,
  SharedWorkerConnection: () => import_chunk_GXHU364O.SharedWorkerConnection,
  Shell: () => import_chunk_GXHU364O.Shell,
  ShellManager: () => import_chunk_GXHU364O.ShellManager,
  WorkerClientServices: () => import_chunk_GXHU364O.WorkerClientServices,
  createClientServices: () => import_chunk_GXHU364O.createClientServices,
  fromAgent: () => import_chunk_GXHU364O.fromAgent,
  fromHost: () => import_chunk_GXHU364O.fromHost,
  fromSocket: () => import_chunk_GXHU364O.fromSocket,
  fromWorker: () => import_chunk_GXHU364O.fromWorker,
  getUnixSocket: () => import_chunk_GXHU364O.getUnixSocket
});
module.exports = __toCommonJS(services_NYQN2ZCW_exports);
var import_chunk_GXHU364O = require("./chunk-GXHU364O.cjs");
var import_chunk_MVOI3MV3 = require("./chunk-MVOI3MV3.cjs");
var import_chunk_P7Y3AD6I = require("./chunk-P7Y3AD6I.cjs");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AgentClientServiceProvider,
  AgentManagerClient,
  ClientServicesProxy,
  FakeAgentHostingProvider,
  IFrameManager,
  LocalClientServices,
  SharedWorkerConnection,
  Shell,
  ShellManager,
  WorkerClientServices,
  createClientServices,
  fromAgent,
  fromHost,
  fromSocket,
  fromWorker,
  getUnixSocket
});
