// @dxos/client/dist/lib/node/chunk-Y34HQJFZ.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var chunk_Y34HQJFZ_exports = {};
__export(chunk_Y34HQJFZ_exports, {
  SpaceProxy: () => SpaceProxy
});
module.exports = __toCommonJS(chunk_Y34HQJFZ_exports);
var import_chunk_MOPAUFKF = require("./chunk-MOPAUFKF.cjs");
var import_chunk_P7Y3AD6I = require("./chunk-P7Y3AD6I.cjs");
var import_lodash = __toESM(require("lodash.isequalwith"));
var import_async = require("@dxos/async");
var import_client_protocol = require("@dxos/client-protocol");
var import_stream = require("@dxos/codec-protobuf/stream");
var import_context = require("@dxos/context");
var import_credentials = require("@dxos/credentials");
var import_debug = require("@dxos/debug");
var import_echo_db = require("@dxos/echo-db");
var import_invariant = require("@dxos/invariant");
var import_log = require("@dxos/log");
var import_protocols = require("@dxos/protocols");
var import_services = require("@dxos/protocols/proto/dxos/client/services");
var import_filter = require("@dxos/protocols/proto/dxos/echo/filter");
var import_credentials2 = require("@dxos/protocols/proto/dxos/halo/credentials");
var import_timeframe = require("@dxos/timeframe");
var import_tracing = require("@dxos/tracing");
function _ts_decorate(decorators, target, key, desc) {
  var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
  if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
  else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
  return c > 3 && r && Object.defineProperty(target, key, r), r;
}
var __dxlog_file = "/home/runner/work/dxos/dxos/packages/sdk/client/src/echo/space-proxy.ts";
var EPOCH_CREATION_TIMEOUT = 6e4;
var _a;
var SpaceProxy = class {
  constructor(_clientServices, _data, echoClient) {
    this._clientServices = _clientServices;
    this._data = _data;
    this._ctx = new import_context.Context(void 0, {
      F: __dxlog_file,
      L: 63
    });
    this._anySpaceUpdate = new import_async.Event();
    this._stateUpdate = new import_async.Event();
    this._pipelineUpdate = new import_async.Event();
    this._databaseInitialized = new import_async.Trigger();
    this._initializationComplete = new import_async.Trigger();
    this._initializing = false;
    this._initialized = false;
    this._state = import_async.MulticastObservable.from(this._stateUpdate, import_services.SpaceState.SPACE_CLOSED);
    this._pipeline = import_async.MulticastObservable.from(this._pipelineUpdate, {});
    this._membersUpdate = new import_async.Event();
    this._members = import_async.MulticastObservable.from(this._membersUpdate, []);
    this._databaseOpen = false;
    this._error = void 0;
    this._properties = void 0;
    this[_inspectCustom] = (depth, options, inspect) => {
      return `${options.stylize(this[Symbol.toStringTag], "special")} ${inspect({
        id: this.id,
        state: import_services.SpaceState[this.state.get()]
      })}`;
    };
    (0, import_log.log)("construct", {
      key: _data.spaceKey,
      state: import_services.SpaceState[_data.state]
    }, {
      F: __dxlog_file,
      L: 122,
      S: this,
      C: (f, a) => f(...a)
    });
    (0, import_invariant.invariant)(this._clientServices.services.InvitationsService, "InvitationsService not available", {
      F: __dxlog_file,
      L: 123,
      S: this,
      A: [
        "this._clientServices.services.InvitationsService",
        "'InvitationsService not available'"
      ]
    });
    this._invitationsProxy = new import_chunk_MOPAUFKF.InvitationsProxy(this._clientServices.services.InvitationsService, this._clientServices.services.IdentityService, () => ({
      kind: import_services.Invitation.Kind.SPACE,
      spaceKey: this.key
    }));
    this._db = echoClient.constructDatabase({
      spaceId: this.id,
      spaceKey: this.key,
      owningObject: this
    });
    this._queues = echoClient.constructQueueFactory(this.id);
    const self = this;
    this._internal = {
      get data() {
        return self._data;
      },
      createEpoch: this._createEpoch.bind(this),
      getCredentials: this._getCredentials.bind(this),
      getEpochs: this._getEpochs.bind(this),
      removeMember: this._removeMember.bind(this),
      migrate: this._migrate.bind(this),
      setEdgeReplicationPreference: this._setEdgeReplicationPreference.bind(this),
      export: this._export.bind(this)
    };
    this._error = this._data.error ? (0, import_protocols.decodeError)(this._data.error) : void 0;
    this._stateUpdate.emit(this._currentState);
    this._pipelineUpdate.emit(_data.pipeline ?? {});
    this._membersUpdate.emit(_data.members ?? []);
  }
  toJSON() {
    return {
      id: this.id,
      db: this._db.toJSON(),
      state: import_services.SpaceState[this.state.get()]
    };
  }
  get id() {
    return this._data.id;
  }
  get key() {
    return this._data.spaceKey;
  }
  get db() {
    return this._db;
  }
  get queues() {
    return this._queues;
  }
  /**
  * @deprecated
  */
  get crud() {
    return this._db.coreDatabase;
  }
  get isOpen() {
    return this._data.state === import_services.SpaceState.SPACE_READY && this._initialized;
  }
  get properties() {
    this._throwIfNotInitialized();
    (0, import_invariant.invariant)(this._properties, "Properties not available", {
      F: __dxlog_file,
      L: 199,
      S: this,
      A: [
        "this._properties",
        "'Properties not available'"
      ]
    });
    return this._properties;
  }
  get state() {
    return this._state;
  }
  /**
  * @inheritdoc
  */
  get pipeline() {
    return this._pipeline;
  }
  /**
  * @inheritdoc
  */
  get invitations() {
    return this._invitationsProxy.created;
  }
  /**
  * @inheritdoc
  */
  get members() {
    return this._members;
  }
  /**
  * @inheritdoc
  */
  // TODO(burdon): Remove?
  get internal() {
    return this._internal;
  }
  get error() {
    return this._error;
  }
  get [(_a = Symbol.toStringTag, _inspectCustom = import_debug.inspectCustom, _a)]() {
    return "SpaceProxy";
  }
  /**
  * Current state of the space.
  * The database is ready to be used in `SpaceState.SPACE_READY` state.
  * Presence is available in `SpaceState.SPACE_CONTROL_ONLY` state.
  */
  get _currentState() {
    if (this._data.state === import_services.SpaceState.SPACE_READY && !this._initialized) {
      return import_services.SpaceState.SPACE_INITIALIZING;
    } else {
      return this._data.state;
    }
  }
  /**
  * Called by EchoProxy to update this space instance.
  * Called once on initial creation.
  * @internal Package-private.
  */
  async _processSpaceUpdate(space) {
    const emitEvent = shouldUpdate(this._data, space);
    const emitPipelineEvent = shouldPipelineUpdate(this._data, space);
    const emitMembersEvent = shouldMembersUpdate(this._data.members, space.members);
    const isFirstTimeInitializing = space.state === import_services.SpaceState.SPACE_READY && !(this._initialized || this._initializing);
    const isReopening = this._data.state !== import_services.SpaceState.SPACE_READY && space.state === import_services.SpaceState.SPACE_READY && !this._databaseOpen;
    const shouldReset = this._databaseOpen && space.state === import_services.SpaceState.SPACE_REQUIRES_MIGRATION;
    (0, import_log.log)("update", {
      key: space.spaceKey,
      prevState: import_services.SpaceState[this._data.state],
      state: import_services.SpaceState[space.state],
      emitEvent,
      emitPipelineEvent,
      emitMembersEvent,
      isFirstTimeInitializing,
      isReopening
    }, {
      F: __dxlog_file,
      L: 281,
      S: this,
      C: (f, a) => f(...a)
    });
    this._data = space;
    if (isFirstTimeInitializing) {
      await this._initialize();
    } else if (isReopening) {
      await this._initializeDb();
    } else if (shouldReset) {
      await this._reset();
    }
    if (space.error) {
      this._error = (0, import_protocols.decodeError)(space.error);
    }
    if (this._initialized) {
      const automergeRoot = space.pipeline?.spaceRootUrl;
      if (automergeRoot) {
        (0, import_log.log)("set space root", {
          spaceKey: this.key,
          automergeRoot
        }, {
          F: __dxlog_file,
          L: 310,
          S: this,
          C: (f, a) => f(...a)
        });
        await this._db.setSpaceRoot(automergeRoot);
      }
    }
    this._anySpaceUpdate.emit(space);
    if (emitEvent) {
      this._stateUpdate.emit(this._currentState);
    }
    if (emitPipelineEvent) {
      this._pipelineUpdate.emit(space.pipeline ?? {});
    }
    if (emitMembersEvent) {
      this._membersUpdate.emit(space.members);
    }
  }
  async _initialize() {
    if (this._initializing || this._initialized) {
      return;
    }
    (0, import_log.log)("initializing...", {
      space: this.key
    }, {
      F: __dxlog_file,
      L: 333,
      S: this,
      C: (f, a) => f(...a)
    });
    this._initializing = true;
    await this._invitationsProxy.open();
    await this._initializeDb();
    this._initialized = true;
    this._initializing = false;
    this._initializationComplete.wake();
    this._stateUpdate.emit(this._currentState);
    this._data.members && this._membersUpdate.emit(this._data.members);
    (0, import_log.log)("initialized", {
      space: this.key
    }, {
      F: __dxlog_file,
      L: 343,
      S: this,
      C: (f, a) => f(...a)
    });
  }
  async _initializeDb() {
    this._databaseOpen = true;
    {
      const automergeRoot = this._data.pipeline?.spaceRootUrl;
      if (automergeRoot !== void 0) {
        await this._db.setSpaceRoot(automergeRoot);
      } else {
        import_log.log.warn("no automerge root found for space", {
          spaceId: this.id
        }, {
          F: __dxlog_file,
          L: 355,
          S: this,
          C: (f, a) => f(...a)
        });
      }
      await this._db.open();
    }
    (0, import_log.log)("ready", void 0, {
      F: __dxlog_file,
      L: 360,
      S: this,
      C: (f, a) => f(...a)
    });
    this._databaseInitialized.wake();
    const propertiesAvailable = new import_async.Trigger();
    {
      const unsubscribe = this._db.query(import_echo_db.Filter.type(import_client_protocol.PropertiesType), {
        dataLocation: import_filter.QueryOptions.DataLocation.LOCAL
      }).subscribe((query) => {
        if (query.objects.length === 1) {
          this._properties = query.objects[0];
          propertiesAvailable.wake();
          this._stateUpdate.emit(this._currentState);
          (0, import_async.scheduleMicroTask)(this._ctx, () => {
            unsubscribe();
          });
        }
      }, {
        fire: true
      });
    }
    await (0, import_debug.warnAfterTimeout)(5e3, "Finding properties for a space", () => (0, import_context.cancelWithContext)(this._ctx, propertiesAvailable.wait()));
  }
  /**
  * Called by EchoProxy close.
  * @internal Package-private.
  */
  async _destroy() {
    await this._reset();
  }
  async _reset() {
    (0, import_log.log)("destroying...", void 0, {
      F: __dxlog_file,
      L: 401,
      S: this,
      C: (f, a) => f(...a)
    });
    await this._ctx.dispose();
    this._ctx = new import_context.Context(void 0, {
      F: __dxlog_file,
      L: 403
    });
    await this._invitationsProxy.close();
    await this._db.close();
    this._initializationComplete.reset();
    this._databaseInitialized.reset();
    this._initializing = false;
    this._initialized = false;
    this._databaseOpen = false;
    (0, import_log.log)("destroyed", void 0, {
      F: __dxlog_file,
      L: 411,
      S: this,
      C: (f, a) => f(...a)
    });
  }
  async open() {
    await this._clientServices.services.SpacesService.updateSpace({
      spaceKey: this.key,
      state: import_services.SpaceState.SPACE_ACTIVE
    }, {
      timeout: import_chunk_P7Y3AD6I.RPC_TIMEOUT
    });
  }
  async close() {
    if (this._databaseOpen) {
      await this._db.flush();
    }
    await this._clientServices.services.SpacesService.updateSpace({
      spaceKey: this.key,
      state: import_services.SpaceState.SPACE_INACTIVE
    }, {
      timeout: import_chunk_P7Y3AD6I.RPC_TIMEOUT
    });
  }
  /**
  * Waits until the space is in the ready state, with database initialized.
  */
  async waitUntilReady() {
    await (0, import_context.cancelWithContext)(this._ctx, this._initializationComplete.wait());
    return this;
  }
  /**
  * Post a message to the space.
  */
  async postMessage(channel, message) {
    (0, import_invariant.invariant)(this._clientServices.services.SpacesService, "SpacesService not available", {
      F: __dxlog_file,
      L: 443,
      S: this,
      A: [
        "this._clientServices.services.SpacesService",
        "'SpacesService not available'"
      ]
    });
    await this._clientServices.services.SpacesService.postMessage({
      spaceKey: this.key,
      channel,
      message: {
        ...message,
        "@type": message["@type"] || "google.protobuf.Struct"
      }
    }, {
      timeout: import_chunk_P7Y3AD6I.RPC_TIMEOUT
    });
  }
  /**
  * Listen for messages posted to the space.
  */
  listen(channel, callback) {
    (0, import_invariant.invariant)(this._clientServices.services.SpacesService, "SpacesService not available", {
      F: __dxlog_file,
      L: 458,
      S: this,
      A: [
        "this._clientServices.services.SpacesService",
        "'SpacesService not available'"
      ]
    });
    const stream = this._clientServices.services.SpacesService.subscribeMessages({
      spaceKey: this.key,
      channel
    }, {
      timeout: import_chunk_P7Y3AD6I.RPC_TIMEOUT
    });
    stream.subscribe(callback);
    return () => stream.close();
  }
  /**
  * Creates a delegated or interactive invitation.
  */
  share(options) {
    this._throwIfNotInitialized();
    (0, import_log.log)("create invitation", options, {
      F: __dxlog_file,
      L: 472,
      S: this,
      C: (f, a) => f(...a)
    });
    return this._invitationsProxy.share({
      ...options,
      spaceKey: this.key
    });
  }
  async admitContact(contact) {
    await this._clientServices.services.SpacesService.admitContact({
      spaceKey: this.key,
      role: import_credentials2.SpaceMember.Role.ADMIN,
      contact
    });
  }
  /**
  * Requests member role update.
  */
  updateMemberRole(request) {
    this._throwIfNotInitialized();
    return this._clientServices.services.SpacesService.updateMemberRole({
      spaceKey: this.key,
      memberKey: request.memberKey,
      newRole: request.newRole
    });
  }
  /**
  * Implementation method.
  */
  createSnapshot() {
    return (0, import_debug.todo)();
  }
  async _removeMember(memberKey) {
    return this._clientServices.services.SpacesService.updateMemberRole({
      spaceKey: this.key,
      memberKey,
      newRole: import_credentials2.SpaceMember.Role.REMOVED
    });
  }
  async _createEpoch({ migration, automergeRootUrl } = {}) {
    (0, import_log.log)("create epoch", {
      migration,
      automergeRootUrl
    }, {
      F: __dxlog_file,
      L: 516,
      S: this,
      C: (f, a) => f(...a)
    });
    const { controlTimeframe: targetTimeframe } = await this._clientServices.services.SpacesService.createEpoch({
      spaceKey: this.key,
      migration,
      automergeRootUrl
    }, {
      timeout: EPOCH_CREATION_TIMEOUT
    });
    if (targetTimeframe) {
      await (0, import_debug.warnAfterTimeout)(5e3, "Waiting for the created epoch to be applied", () => this._anySpaceUpdate.waitForCondition(() => {
        const currentTimeframe = this._data.pipeline?.currentControlTimeframe;
        return (currentTimeframe && import_timeframe.Timeframe.dependencies(targetTimeframe, currentTimeframe).isEmpty()) ?? false;
      }));
    }
  }
  async _getCredentials() {
    const stream = this._clientServices.services.SpacesService?.queryCredentials({
      spaceKey: this.key,
      noTail: true
    });
    (0, import_invariant.invariant)(stream, "SpacesService not available", {
      F: __dxlog_file,
      L: 538,
      S: this,
      A: [
        "stream",
        "'SpacesService not available'"
      ]
    });
    return await import_stream.Stream.consumeData(stream);
  }
  async _getEpochs() {
    const credentials = await this._getCredentials();
    return credentials.filter((credential) => (0, import_credentials.checkCredentialType)(credential, "dxos.halo.credentials.Epoch"));
  }
  async _migrate() {
    await this._createEpoch({
      migration: import_services.CreateEpochRequest.Migration.MIGRATE_REFERENCES_TO_DXN
    });
    await this._databaseInitialized.wait();
    if (this._db.coreDatabase.getNumberOfInlineObjects() > 1) {
      await this._createEpoch({
        migration: import_services.CreateEpochRequest.Migration.FRAGMENT_AUTOMERGE_ROOT
      });
    }
  }
  async _setEdgeReplicationPreference(setting) {
    await this._clientServices.services.SpacesService.updateSpace({
      spaceKey: this.key,
      edgeReplication: setting
    }, {
      timeout: import_chunk_P7Y3AD6I.RPC_TIMEOUT
    });
  }
  _throwIfNotInitialized() {
    if (!this._initialized) {
      throw new Error("Space is not initialized.");
    }
  }
  async _export() {
    const { archive } = await this._clientServices.services.SpacesService.exportSpace({
      spaceId: this.id
    });
    return archive;
  }
};
_ts_decorate([
  import_tracing.trace.info()
], SpaceProxy.prototype, "_initializing", void 0);
_ts_decorate([
  import_tracing.trace.info()
], SpaceProxy.prototype, "_initialized", void 0);
_ts_decorate([
  import_tracing.trace.info()
], SpaceProxy.prototype, "key", null);
_ts_decorate([
  import_tracing.trace.info()
], SpaceProxy.prototype, "isOpen", null);
_ts_decorate([
  import_tracing.trace.info({
    depth: 2
  })
], SpaceProxy.prototype, "properties", null);
_ts_decorate([
  import_tracing.trace.info({
    enum: import_services.SpaceState
  })
], SpaceProxy.prototype, "_currentState", null);
_ts_decorate([
  import_async.synchronized
], SpaceProxy.prototype, "_processSpaceUpdate", null);
_ts_decorate([
  import_tracing.trace.span({
    showInBrowserTimeline: true
  })
], SpaceProxy.prototype, "_initializeDb", null);
_ts_decorate([
  import_async.synchronized
], SpaceProxy.prototype, "_destroy", null);
SpaceProxy = _ts_decorate([
  import_tracing.trace.resource()
], SpaceProxy);
var shouldUpdate = (prev, next) => {
  return prev.state !== next.state;
};
var shouldPipelineUpdate = (prev, next) => {
  return !(0, import_lodash.default)(prev.pipeline, next.pipeline, import_debug.loadashEqualityFn);
};
var shouldMembersUpdate = (prev, next) => {
  if (!next) {
    return false;
  }
  return !(0, import_lodash.default)(prev, next, import_debug.loadashEqualityFn);
};
var _inspectCustom;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SpaceProxy
});
