// @dxos/echo/dist/lib/node/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var node_exports = {};
__export(node_exports, {
  DXN: () => import_chunk_4HQE2F3L.DXN,
  Filter: () => import_chunk_4HQE2F3L.Filter,
  Key: () => import_chunk_4HQE2F3L.Key_exports,
  Obj: () => import_chunk_4HQE2F3L.Obj_exports,
  Query: () => import_chunk_4HQE2F3L.Query,
  Ref: () => import_chunk_4HQE2F3L.Ref_exports,
  Relation: () => import_chunk_4HQE2F3L.Relation_exports,
  Type: () => import_chunk_4HQE2F3L.Type_exports
});
module.exports = __toCommonJS(node_exports);
var import_chunk_4HQE2F3L = require("./chunk-4HQE2F3L.cjs");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DXN,
  Filter,
  Key,
  Obj,
  Query,
  Ref,
  Relation,
  Type
});
