// @dxos/protocols/dist/cjs/src/proto/gen/dxos/edge/messenger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SwarmRequest = exports.ConnectionState = void 0;
/**
 * Defined in:
 *   {@link file://./../../../dxos/edge/messenger.proto}
 */
var ConnectionState;
(function (ConnectionState) {
    ConnectionState[ConnectionState["CONNECTED"] = 0] = "CONNECTED";
    ConnectionState[ConnectionState["DISCONNECTED_NORMAL"] = 1] = "DISCONNECTED_NORMAL";
    ConnectionState[ConnectionState["DISCONNECTED_ABRUPT"] = 2] = "DISCONNECTED_ABRUPT";
})(ConnectionState || (exports.ConnectionState = ConnectionState = {}));
var SwarmRequest;
(function (SwarmRequest) {
    /**
     * Defined in:
     *   {@link file://./../../../dxos/edge/messenger.proto}
     */
    let Action;
    (function (Action) {
        Action[Action["INFO"] = 0] = "INFO";
        Action[Action["JOIN"] = 1] = "JOIN";
        Action[Action["LEAVE"] = 2] = "LEAVE";
    })(Action = SwarmRequest.Action || (SwarmRequest.Action = {}));
})(SwarmRequest || (exports.SwarmRequest = SwarmRequest = {}));
