// @dxos/network-manager/dist/lib/node/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var node_exports = {};
__export(node_exports, {
  Connection: () => import_chunk_2G6RZMS5.Connection,
  ConnectionLimiter: () => import_chunk_2G6RZMS5.ConnectionLimiter,
  ConnectionLog: () => import_chunk_2G6RZMS5.ConnectionLog,
  ConnectionState: () => import_chunk_2G6RZMS5.ConnectionState,
  EventType: () => import_chunk_2G6RZMS5.EventType,
  FullyConnectedTopology: () => import_chunk_2G6RZMS5.FullyConnectedTopology,
  MAX_CONCURRENT_INITIATING_CONNECTIONS: () => import_chunk_2G6RZMS5.MAX_CONCURRENT_INITIATING_CONNECTIONS,
  MMSTTopology: () => import_chunk_2G6RZMS5.MMSTTopology,
  MemoryTransport: () => import_chunk_2G6RZMS5.MemoryTransport,
  MemoryTransportFactory: () => import_chunk_2G6RZMS5.MemoryTransportFactory,
  RtcTransportProxy: () => import_chunk_2G6RZMS5.RtcTransportProxy,
  RtcTransportProxyFactory: () => import_chunk_2G6RZMS5.RtcTransportProxyFactory,
  RtcTransportService: () => import_chunk_2G6RZMS5.RtcTransportService,
  StarTopology: () => import_chunk_2G6RZMS5.StarTopology,
  Swarm: () => import_chunk_2G6RZMS5.Swarm,
  SwarmMapper: () => import_chunk_2G6RZMS5.SwarmMapper,
  SwarmMessenger: () => import_chunk_2G6RZMS5.SwarmMessenger,
  SwarmNetworkManager: () => import_chunk_2G6RZMS5.SwarmNetworkManager,
  TransportKind: () => import_chunk_2G6RZMS5.TransportKind,
  createIceProvider: () => import_chunk_2G6RZMS5.createIceProvider,
  createRtcTransportFactory: () => import_chunk_2G6RZMS5.createRtcTransportFactory,
  createTeleportProtocolFactory: () => import_chunk_2G6RZMS5.createTeleportProtocolFactory
});
module.exports = __toCommonJS(node_exports);
var import_chunk_2G6RZMS5 = require("./chunk-2G6RZMS5.cjs");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Connection,
  ConnectionLimiter,
  ConnectionLog,
  ConnectionState,
  EventType,
  FullyConnectedTopology,
  MAX_CONCURRENT_INITIATING_CONNECTIONS,
  MMSTTopology,
  MemoryTransport,
  MemoryTransportFactory,
  RtcTransportProxy,
  RtcTransportProxyFactory,
  RtcTransportService,
  StarTopology,
  Swarm,
  SwarmMapper,
  SwarmMessenger,
  SwarmNetworkManager,
  TransportKind,
  createIceProvider,
  createRtcTransportFactory,
  createTeleportProtocolFactory
});
