// @dxos/protocols/src/buf/proto/gen/dxos/mesh/protocol_pb.ts
//
// Copyright 2021 DXOS.org
//

// @generated by protoc-gen-es v2.1.0 with parameter "target=ts,import_extension=js"
// @generated from file dxos/mesh/protocol.proto (package dxos.mesh.protocol, syntax proto3)
/* eslint-disable */

import type { GenFile, GenMessage } from "@bufbuild/protobuf/codegenv1";
import { fileDesc, messageDesc } from "@bufbuild/protobuf/codegenv1";
import type { Any } from "@bufbuild/protobuf/wkt";
import { file_google_protobuf_any } from "@bufbuild/protobuf/wkt";
import type { Message as Message$1 } from "@bufbuild/protobuf";

/**
 * Describes the file dxos/mesh/protocol.proto.
 */
export const file_dxos_mesh_protocol: GenFile = /*@__PURE__*/
  fileDesc("ChhkeG9zL21lc2gvcHJvdG9jb2wucHJvdG8SEmR4b3MubWVzaC5wcm90b2NvbCJmCgdNZXNzYWdlEgwKBG5tSWQYASABKAkSEgoKbm1SZXNwb25zZRgCIAEoCBITCgtubUVwaGVtZXJhbBgDIAEoCBIkCgZubURhdGEYBCABKAsyFC5nb29nbGUucHJvdG9idWYuQW55IiYKBUVycm9yEgwKBGNvZGUYASABKAkSDwoHbWVzc2FnZRgCIAEoCSIWCgZCdWZmZXISDAoEZGF0YRgBIAEoDGIGcHJvdG8z", [file_google_protobuf_any]);

/**
 * TODO(burdon): Rename ProtoMessage.
 *
 * @generated from message dxos.mesh.protocol.Message
 */
export type Message = Message$1<"dxos.mesh.protocol.Message"> & {
  /**
   * TODO(burdon): nm???
   *
   * @generated from field: string nmId = 1;
   */
  nmId: string;

  /**
   * @generated from field: bool nmResponse = 2;
   */
  nmResponse: boolean;

  /**
   * @generated from field: bool nmEphemeral = 3;
   */
  nmEphemeral: boolean;

  /**
   * @generated from field: google.protobuf.Any nmData = 4;
   */
  nmData?: Any;
};

/**
 * Describes the message dxos.mesh.protocol.Message.
 * Use `create(MessageSchema)` to create a new message.
 */
export const MessageSchema: GenMessage<Message> = /*@__PURE__*/
  messageDesc(file_dxos_mesh_protocol, 0);

/**
 * TODO(burdon): Rename ProtoError.
 *
 * @generated from message dxos.mesh.protocol.Error
 */
export type Error = Message$1<"dxos.mesh.protocol.Error"> & {
  /**
   * @generated from field: string code = 1;
   */
  code: string;

  /**
   * @generated from field: string message = 2;
   */
  message: string;
};

/**
 * Describes the message dxos.mesh.protocol.Error.
 * Use `create(ErrorSchema)` to create a new message.
 */
export const ErrorSchema: GenMessage<Error> = /*@__PURE__*/
  messageDesc(file_dxos_mesh_protocol, 1);

/**
 * TODO(burdon): Rename ProtoBuffer.
 *
 * @generated from message dxos.mesh.protocol.Buffer
 */
export type Buffer = Message$1<"dxos.mesh.protocol.Buffer"> & {
  /**
   * @generated from field: bytes data = 1;
   */
  data: Uint8Array;
};

/**
 * Describes the message dxos.mesh.protocol.Buffer.
 * Use `create(BufferSchema)` to create a new message.
 */
export const BufferSchema: GenMessage<Buffer> = /*@__PURE__*/
  messageDesc(file_dxos_mesh_protocol, 2);
