// @dxos/protocols/dist/cjs/src/tracing.js
"use strict";
//
// Copyright 2023 DXOS.org
//
Object.defineProperty(exports, "__esModule", { value: true });
exports.trace = void 0;
exports.trace = {
    begin: ({ id, parentId, data }) => {
        return {
            span: {
                command: 'begin',
                id,
                parent: parentId,
                data,
            },
        };
    },
    end: ({ id, status, data }) => {
        return {
            span: {
                command: 'end',
                id,
                status: status !== null && status !== void 0 ? status : 'ok',
                data,
            },
        };
    },
    update: ({ id, data }) => {
        return {
            span: {
                command: 'update',
                id,
                data,
            },
        };
    },
    error: ({ id, error, data }) => {
        return {
            span: {
                command: 'end',
                id,
                status: 'error',
                data: {
                    errorMessage: error.message,
                    errorName: error.name,
                    errorStack: error.stack,
                    ...data,
                },
            },
        };
    },
};
