// @dxos/react-ui/dist/lib/node/testing/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var testing_exports = {};
__export(testing_exports, {
  withSurfaceVariantsLayout: () => withSurfaceVariantsLayout,
  withTheme: () => withTheme
});
module.exports = __toCommonJS(testing_exports);
var import_chunk_3E3QKWX4 = require("../chunk-3E3QKWX4.cjs");
var import_react = require("react");
var import_react_ui_theme = require("@dxos/react-ui-theme");
var import_tracking = require("@preact-signals/safe-react/tracking");
var import_react2 = __toESM(require("react"));
var import_react_ui_theme2 = require("@dxos/react-ui-theme");
var withTheme = (Story, context) => {
  const theme = context?.parameters?.theme || context?.globals?.theme;
  (0, import_react.useEffect)(() => {
    document.documentElement.classList[theme === "dark" ? "add" : "remove"]("dark");
  }, [
    theme
  ]);
  return (0, import_react.createElement)(import_chunk_3E3QKWX4.ThemeProvider, {
    children: (0, import_react.createElement)(Story),
    tx: import_react_ui_theme.defaultTx
  });
};
var Container = ({ children, elevation, surface }) => {
  var _effect = (0, import_tracking.useSignals)();
  try {
    return /* @__PURE__ */ import_react2.default.createElement("div", {
      className: (0, import_react_ui_theme2.mx)("rounded-md border border-separator", surface, (0, import_react_ui_theme2.surfaceShadow)({
        elevation
      }))
    }, children);
  } finally {
    _effect.f();
  }
};
var Panel = ({ Story, elevations, densities, className }) => {
  var _effect = (0, import_tracking.useSignals)();
  try {
    return /* @__PURE__ */ import_react2.default.createElement("div", {
      className: (0, import_react_ui_theme2.mx)("flex flex-col h-full p-4 gap-4", className)
    }, elevations?.map(({ elevation, surface }) => densities?.map((density) => /* @__PURE__ */ import_react2.default.createElement(Container, {
      key: `${elevation}--${density}`,
      surface,
      elevation
    }, /* @__PURE__ */ import_react2.default.createElement(Story, null)))));
  } finally {
    _effect.f();
  }
};
var withSurfaceVariantsLayout = ({ elevations = [
  {
    elevation: "base",
    surface: "bg-baseSurface"
  },
  {
    elevation: "positioned",
    surface: "bg-cardSurface"
  },
  {
    elevation: "dialog",
    surface: "bg-modalSurface"
  }
], densities = [
  "coarse"
] } = {}) => {
  return (Story) => /* @__PURE__ */ import_react2.default.createElement("div", {
    className: "fixed inset-0 grid grid-cols-2 overflow-y-auto"
  }, /* @__PURE__ */ import_react2.default.createElement(Panel, {
    Story,
    className: "light",
    elevations,
    densities
  }), /* @__PURE__ */ import_react2.default.createElement(Panel, {
    Story,
    className: "dark",
    elevations,
    densities
  }));
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  withSurfaceVariantsLayout,
  withTheme
});
