// @dxos/network-manager/dist/lib/node/testing/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var testing_exports = {};
__export(testing_exports, {
  TEST_SIGNAL_HOSTS: () => TEST_SIGNAL_HOSTS,
  TestBuilder: () => TestBuilder,
  TestPeer: () => TestPeer,
  TestSwarmConnection: () => TestSwarmConnection
});
module.exports = __toCommonJS(testing_exports);
var import_chunk_2G6RZMS5 = require("../chunk-2G6RZMS5.cjs");
var import_keys = require("@dxos/keys");
var import_log = require("@dxos/log");
var import_messaging = require("@dxos/messaging");
var import_proto = require("@dxos/protocols/proto");
var import_services = require("@dxos/protocols/proto/dxos/client/services");
var import_rpc = require("@dxos/rpc");
var import_util = require("@dxos/util");
var import_tcp_transport = require("#tcp-transport");
var import_async = require("@dxos/async");
var import_keys2 = require("@dxos/keys");
var import_log2 = require("@dxos/log");
var import_teleport = require("@dxos/teleport");
var import_util2 = require("@dxos/util");
var __dxlog_file = "/home/runner/work/dxos/dxos/packages/core/mesh/network-manager/src/testing/test-wire-protocol.ts";
var TestWireProtocol = class {
  constructor(_extensionFactory = () => []) {
    this._extensionFactory = _extensionFactory;
    this.connections = new import_util2.ComplexMap(import_keys2.PublicKey.hash);
    this.streamConnections = new import_util2.ComplexMap(import_keys2.PublicKey.hash);
    this.connected = new import_async.Event();
    this.disconnected = new import_async.Event();
    this.otherConnections = new import_util2.ComplexMap(({ remotePeerId, extension }) => remotePeerId.toHex() + extension);
    this.factory = (0, import_chunk_2G6RZMS5.createTeleportProtocolFactory)(async (teleport) => {
      (0, import_log2.log)("create", {
        remotePeerId: teleport.remotePeerId
      }, {
        F: __dxlog_file,
        L: 35,
        S: this,
        C: (f, a) => f(...a)
      });
      const handleDisconnect = () => {
        this.connections.delete(teleport.remotePeerId);
        this.disconnected.emit(teleport.remotePeerId);
      };
      const extension = new import_teleport.TestExtension({
        onClose: async () => handleDisconnect(),
        onAbort: async () => handleDisconnect()
      });
      this.connections.set(teleport.remotePeerId, extension);
      teleport.addExtension("test", extension);
      this.connected.emit(teleport.remotePeerId);
      const streamExtension = new import_teleport.TestExtensionWithStreams({
        onClose: async () => {
          this.streamConnections.delete(teleport.remotePeerId);
        }
      });
      this.streamConnections.set(teleport.remotePeerId, streamExtension);
      teleport.addExtension("test-stream", streamExtension);
      for (const { name, extension: extension2 } of this._extensionFactory()) {
        this.otherConnections.set({
          remotePeerId: teleport.remotePeerId,
          extension: name
        }, extension2);
        teleport.addExtension(name, extension2);
      }
    });
  }
  async waitForConnection(peerId) {
    if (this.connections.has(peerId)) {
      return this.connections.get(peerId);
    }
    (0, import_log2.log)("waitForConnection", {
      peerId
    }, {
      F: __dxlog_file,
      L: 66,
      S: this,
      C: (f, a) => f(...a)
    });
    await (0, import_async.asyncTimeout)(
      this.connected.waitFor((connectedId) => connectedId.equals(peerId)),
      // TODO(nf): Make this configurable.
      1e4
    );
    return this.connections.get(peerId);
  }
  async testConnection(peerId, message) {
    const connection = await this.waitForConnection(peerId);
    await connection.test(message);
  }
  async openStream(peerId, streamTag, streamLoadInterval, streamLoadChunkSize) {
    if (!this.streamConnections.has(peerId)) {
      throw new Error("Connection does not exist.");
    }
    const connection = this.streamConnections.get(peerId);
    return connection.addNewStream(streamLoadInterval, streamLoadChunkSize, streamTag);
  }
  async closeStream(peerId, streamTag) {
    if (!this.streamConnections.has(peerId)) {
      throw new Error("Connection does not exist.");
    }
    const connection = this.streamConnections.get(peerId);
    return connection.closeStream(streamTag);
  }
};
var __dxlog_file2 = "/home/runner/work/dxos/dxos/packages/core/mesh/network-manager/src/testing/test-builder.ts";
var port = process.env.SIGNAL_PORT ?? 4e3;
var TEST_SIGNAL_HOSTS = [
  {
    server: `ws://localhost:${port}/.well-known/dx/signal`
  }
];
var TestBuilder = class {
  constructor(options = {}) {
    this.options = options;
    this._signalContext = new import_messaging.MemorySignalManagerContext();
  }
  createSignalManager() {
    if (this.options.signalHosts) {
      return new import_messaging.WebsocketSignalManager(this.options.signalHosts);
    }
    return new import_messaging.MemorySignalManager(this._signalContext);
  }
  createPeer(peerId = import_keys.PublicKey.random()) {
    return new TestPeer(this, peerId, this.options.transport);
  }
};
var TestPeer = class {
  constructor(testBuilder, peerId, transport = testBuilder.options.signalHosts ? import_chunk_2G6RZMS5.TransportKind.WEB_RTC : import_chunk_2G6RZMS5.TransportKind.MEMORY) {
    this.testBuilder = testBuilder;
    this.peerId = peerId;
    this.transport = transport;
    this._swarms = new import_util.ComplexMap(import_keys.PublicKey.hash);
    this._signalManager = this.testBuilder.createSignalManager();
    this._networkManager = this.createNetworkManager(this.transport);
    this._networkManager.setPeerInfo({
      identityKey: peerId.toHex(),
      peerKey: peerId.toHex()
    });
  }
  // TODO(burdon): Move to TestBuilder.
  createNetworkManager(transport) {
    let transportFactory;
    if (this.testBuilder.options.signalHosts) {
      import_log.log.info(`using ${transport} transport with signal server.`, void 0, {
        F: __dxlog_file2,
        L: 94,
        S: this,
        C: (f, a) => f(...a)
      });
      switch (transport) {
        case import_chunk_2G6RZMS5.TransportKind.MEMORY:
          throw new Error("Memory transport not supported with signal server.");
        case import_chunk_2G6RZMS5.TransportKind.TCP:
          transportFactory = import_tcp_transport.TcpTransportFactory;
          break;
        case import_chunk_2G6RZMS5.TransportKind.WEB_RTC:
          transportFactory = (0, import_chunk_2G6RZMS5.createRtcTransportFactory)();
          break;
        case import_chunk_2G6RZMS5.TransportKind.WEB_RTC_PROXY:
          {
            const [proxyPort, servicePort] = (0, import_rpc.createLinkedPorts)();
            this._proxy = (0, import_rpc.createProtoRpcPeer)({
              port: proxyPort,
              requested: {
                BridgeService: import_proto.schema.getService("dxos.mesh.bridge.BridgeService")
              },
              noHandshake: true,
              encodingOptions: {
                preserveAny: true
              }
            });
            this._service = (0, import_rpc.createProtoRpcPeer)({
              port: servicePort,
              exposed: {
                BridgeService: import_proto.schema.getService("dxos.mesh.bridge.BridgeService")
              },
              handlers: {
                BridgeService: new import_chunk_2G6RZMS5.RtcTransportService()
              },
              noHandshake: true,
              encodingOptions: {
                preserveAny: true
              }
            });
            transportFactory = new import_chunk_2G6RZMS5.RtcTransportProxyFactory().setBridgeService(this._proxy.rpc.BridgeService);
          }
          break;
        default:
          throw new Error(`Unsupported transport: ${transport}`);
      }
    } else {
      if (transport !== import_chunk_2G6RZMS5.TransportKind.MEMORY && transport !== import_chunk_2G6RZMS5.TransportKind.TCP) {
        import_log.log.warn(`specified transport ${transport} but no signalling configured, using memory transport instead`, void 0, {
          F: __dxlog_file2,
          L: 140,
          S: this,
          C: (f, a) => f(...a)
        });
      }
      import_log.log.info(`using ${transport} transport without signal server.`, void 0, {
        F: __dxlog_file2,
        L: 142,
        S: this,
        C: (f, a) => f(...a)
      });
      transportFactory = import_chunk_2G6RZMS5.MemoryTransportFactory;
    }
    return new import_chunk_2G6RZMS5.SwarmNetworkManager({
      signalManager: this._signalManager,
      transportFactory
    });
  }
  async open() {
    await this._networkManager.open();
    await this._proxy?.open();
    await this._service?.open();
  }
  async close() {
    await Promise.all(Array.from(this._swarms.values()).map((swarm) => swarm.leave()));
    this._swarms.clear();
    await this._proxy?.close();
    await this._service?.close();
    await this._networkManager.close();
  }
  getSwarm(topic) {
    const swarm = this._swarms.get(topic);
    if (!swarm) {
      throw new Error(`Swarm not found for topic: ${topic}`);
    }
    return swarm;
  }
  createSwarm(topic, extensionFactory = () => []) {
    const swarm = new TestSwarmConnection(this, topic, extensionFactory);
    this._swarms.set(topic, swarm);
    return swarm;
  }
  async goOffline() {
    await this._networkManager.setConnectionState(import_services.ConnectionState.OFFLINE);
  }
  async goOnline() {
    await this._networkManager.setConnectionState(import_services.ConnectionState.ONLINE);
  }
};
var TestSwarmConnection = class {
  constructor(peer, topic, extensionFactory) {
    this.peer = peer;
    this.topic = topic;
    this.extensionFactory = extensionFactory;
    this.protocol = new TestWireProtocol(this.extensionFactory);
  }
  // TODO(burdon): Need to create new plugin instance per swarm?
  //  If so, then perhaps joinSwarm should return swarm object with access to plugins.
  async join(topology = new import_chunk_2G6RZMS5.FullyConnectedTopology()) {
    await this.peer._networkManager.joinSwarm({
      topic: this.topic,
      peerInfo: {
        peerKey: this.peer.peerId.toHex(),
        identityKey: this.peer.peerId.toHex()
      },
      protocolProvider: this.protocol.factory,
      topology
    });
    return this;
  }
  async leave() {
    await this.peer._networkManager.leaveSwarm(this.topic);
    return this;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TEST_SIGNAL_HOSTS,
  TestBuilder,
  TestPeer,
  TestSwarmConnection
});
