// @dxos/log/dist/lib/node/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var node_exports = {};
__export(node_exports, {
  BROWSER_PROCESSOR: () => BROWSER_PROCESSOR,
  CONSOLE_PROCESSOR: () => CONSOLE_PROCESSOR,
  DEBUG_PROCESSOR: () => DEBUG_PROCESSOR,
  DEFAULT_FORMATTER: () => DEFAULT_FORMATTER,
  FILE_PROCESSOR: () => FILE_PROCESSOR,
  LogLevel: () => LogLevel,
  LogProcessorType: () => LogProcessorType,
  SHORT_FORMATTER: () => SHORT_FORMATTER,
  createFileProcessor: () => createFileProcessor,
  debug: () => debug,
  gatherLogInfoFromScope: () => gatherLogInfoFromScope,
  getContextFromEntry: () => getContextFromEntry,
  getCurrentOwnershipScope: () => getCurrentOwnershipScope,
  getRelativeFilename: () => getRelativeFilename,
  levels: () => levels,
  log: () => log,
  logInfo: () => logInfo,
  omit: () => import_lodash.default,
  parseFilter: () => parseFilter,
  pick: () => import_util.pick,
  shortLevelName: () => shortLevelName,
  shouldLog: () => shouldLog,
  truncate: () => truncate
});
module.exports = __toCommonJS(node_exports);
var import_lodash = __toESM(require("lodash.omit"));
var import_util = require("@dxos/util");
var import_chalk = __toESM(require("chalk"));
var import_node_util = require("node:util");
var import_lodash2 = __toESM(require("lodash.defaultsdeep"));
var import_js_yaml = __toESM(require("js-yaml"));
var import_node_fs = __toESM(require("node:fs"));
var import_chalk2 = __toESM(require("chalk"));
var import_node_util2 = require("node:util");
var import_util2 = require("@dxos/util");
var import_node_util3 = require("node:util");
var import_util3 = require("@dxos/util");
var import_node_fs2 = require("node:fs");
var import_node_path = require("node:path");
var import_util4 = require("@dxos/util");
var import_node_util4 = require("node:util");
var LogLevel = /* @__PURE__ */ function(LogLevel2) {
  LogLevel2[LogLevel2["TRACE"] = 5] = "TRACE";
  LogLevel2[LogLevel2["DEBUG"] = 10] = "DEBUG";
  LogLevel2[LogLevel2["VERBOSE"] = 11] = "VERBOSE";
  LogLevel2[LogLevel2["INFO"] = 12] = "INFO";
  LogLevel2[LogLevel2["WARN"] = 13] = "WARN";
  LogLevel2[LogLevel2["ERROR"] = 14] = "ERROR";
  return LogLevel2;
}({});
var levels = {
  trace: 5,
  debug: 10,
  verbose: 11,
  info: 12,
  warn: 13,
  error: 14
};
var shortLevelName = {
  [5]: "T",
  [10]: "D",
  [11]: "V",
  [12]: "I",
  [13]: "W",
  [14]: "E"
};
var LogProcessorType = /* @__PURE__ */ function(LogProcessorType2) {
  LogProcessorType2["CONSOLE"] = "console";
  LogProcessorType2["BROWSER"] = "browser";
  LogProcessorType2["DEBUG"] = "debug";
  return LogProcessorType2;
}({});
var logInfoProperties = Symbol("logInfoProperties");
var logInfo = (target, propertyKey, descriptor) => {
  (target[logInfoProperties] ??= []).push(propertyKey);
};
var gatherLogInfoFromScope = (scope) => {
  if (!scope) {
    return {};
  }
  const res = {};
  const prototype = Object.getPrototypeOf(scope);
  const infoProps = prototype[logInfoProperties] ?? [];
  for (const prop of infoProps) {
    try {
      res[prop] = typeof scope[prop] === "function" ? scope[prop]() : scope[prop];
    } catch (err) {
      res[prop] = err.message;
    }
  }
  return res;
};
var matchFilter = (filter, level, path) => {
  return level >= filter.level && (!filter.pattern || path.includes(filter.pattern));
};
var shouldLog = (entry, filters) => {
  if (filters === void 0) {
    return true;
  } else {
    return filters.some((filter) => matchFilter(filter, entry.level, entry.meta?.F ?? ""));
  }
};
var getContextFromEntry = (entry) => {
  let context;
  if (entry.meta) {
    const scopeInfo = gatherLogInfoFromScope(entry.meta.S);
    if (Object.keys(scopeInfo).length > 0) {
      context = Object.assign(context ?? {}, scopeInfo);
    }
  }
  const entryContext = typeof entry.context === "function" ? entry.context() : entry.context;
  if (entryContext) {
    if (entryContext instanceof Error) {
      const c = entryContext.context;
      context = Object.assign(context ?? {}, {
        error: entryContext.stack,
        ...c
      });
    } else if (typeof entryContext === "object") {
      context = Object.assign(context ?? {}, entryContext);
    }
  }
  if (entry.error) {
    const errorContext = entry.error.context;
    context = Object.assign(context ?? {}, {
      error: entry.error,
      ...errorContext
    });
  }
  return context && Object.keys(context).length > 0 ? context : void 0;
};
var nextPromiseId = 0;
var createMethodLogDecorator = (log2) => (arg0, arg1, meta) => (target, propertyKey, descriptor) => {
  const method = descriptor.value;
  const methodName = propertyKey;
  descriptor.value = function(...args) {
    const combinedMeta = {
      F: "",
      L: 0,
      ...meta ?? {},
      S: this
    };
    const formattedArgs = args.map((arg) => (0, import_node_util.inspect)(arg, false, 1, true)).join(", ");
    try {
      const startTime = performance.now();
      const result = method.apply(this, args);
      if (isThenable(result)) {
        const id = nextPromiseId++;
        logAsyncBegin(log2, methodName, formattedArgs, id, combinedMeta);
        result.then((resolvedValue) => {
          logAsyncResolved(log2, methodName, resolvedValue, id, startTime, combinedMeta);
        }, (err) => {
          logAsyncRejected(log2, methodName, err, id, startTime, combinedMeta);
        });
      } else {
        logSyncCall(log2, methodName, formattedArgs, result, combinedMeta);
      }
      return result;
    } catch (err) {
      logSyncError(log2, methodName, formattedArgs, err, combinedMeta);
      throw err;
    }
  };
  Object.defineProperty(descriptor.value, "name", {
    value: methodName + "$log"
  });
};
var createFunctionLogDecorator = (log2) => (name, fn, opts = {}) => {
  const decoratedFn = function(...args) {
    const combinedMeta = {
      F: "",
      L: 0
    };
    const formattedArgs = args.map((arg) => (0, import_node_util.inspect)(arg, false, 1, true)).join(", ");
    try {
      const startTime = performance.now();
      const result = fn.apply(this, args);
      let transformedResult = result;
      if (opts.transformOutput) {
        if (isThenable(result)) {
          transformedResult = result.then(opts.transformOutput);
        } else {
          transformedResult = opts.transformOutput(result);
        }
      }
      if (isThenable(transformedResult)) {
        const id = nextPromiseId++;
        logAsyncBegin(log2, name, formattedArgs, id, combinedMeta);
        transformedResult.then((resolvedValue) => {
          logAsyncResolved(log2, name, resolvedValue, id, startTime, combinedMeta);
        }, (err) => {
          logAsyncRejected(log2, name, err, id, startTime, combinedMeta);
        });
      } else {
        logSyncCall(log2, name, formattedArgs, transformedResult, combinedMeta);
      }
      return result;
    } catch (err) {
      logSyncError(log2, name, formattedArgs, err, combinedMeta);
      throw err;
    }
  };
  Object.defineProperty(decoratedFn, "name", {
    value: name + "$log"
  });
  return decoratedFn;
};
var isThenable = (obj) => obj && typeof obj.then === "function";
var logSyncCall = (log2, methodName, formattedArgs, result, combinedMeta) => {
  log2.info(`.${formatFunction(methodName)} (${formattedArgs}) ${import_chalk.default.gray("=>")} ${(0, import_node_util.inspect)(result, false, 1, true)}`, {}, combinedMeta);
};
var logSyncError = (log2, methodName, formattedArgs, err, combinedMeta) => {
  log2.error(`.${formatFunction(methodName)} (${formattedArgs}) \u{1F525} ${err}`, {}, combinedMeta);
};
var logAsyncBegin = (log2, methodName, formattedArgs, promiseId, combinedMeta) => {
  log2.info(`.${formatFunction(methodName)} \u21B4 (${formattedArgs}) ${import_chalk.default.gray("=>")} ${formatPromise(promiseId)}`, {}, combinedMeta);
};
var logAsyncResolved = (log2, methodName, resolvedValue, promiseId, startTime, combinedMeta) => {
  if (resolvedValue !== void 0) {
    log2.info(`.${formatFunction(methodName)} \u21B2 ${greenCheck} ${import_chalk.default.gray("resolve")} ${formatPromise(promiseId)} ${formatTimeElapsed(startTime)} ${import_chalk.default.gray("=>")} ${(0, import_node_util.inspect)(resolvedValue, false, 1, true)}`, {}, combinedMeta);
  } else {
    log2.info(`.${formatFunction(methodName)} \u21B2 ${greenCheck} ${import_chalk.default.gray("resolve")} ${formatPromise(promiseId)} ${formatTimeElapsed(startTime)}`, {}, combinedMeta);
  }
};
var logAsyncRejected = (log2, methodName, err, promiseId, startTime, combinedMeta) => {
  log2.info(`.${formatFunction(methodName)} \u21B2 \u{1F525} ${import_chalk.default.gray("reject")} ${formatPromise(promiseId)} ${formatTimeElapsed(startTime)} ${import_chalk.default.gray("=>")} ${err}`, {}, combinedMeta);
};
var greenCheck = typeof import_chalk.default.green === "function" ? import_chalk.default.green("\u2714") : "\u2714";
var formatTimeElapsed = (startTime) => import_chalk.default.gray(`${(performance.now() - startTime).toFixed(0)}ms`);
var COLOR_FUNCTION = [
  220,
  220,
  170
];
var formatFunction = (name) => import_chalk.default.bold(import_chalk.default.rgb(...COLOR_FUNCTION)(name));
var formatPromise = (id) => import_chalk.default.blue(`Promise#${id}`);
var loadOptions = (filepath) => {
  if (filepath) {
    try {
      const text = import_node_fs.default.readFileSync(filepath, "utf-8");
      if (text) {
        return import_js_yaml.default.load(text);
      }
    } catch (err) {
      console.warn(`Invalid log file: ${filepath}`);
    }
  }
};
var getRelativeFilename = (filename) => {
  const match = filename.match(/.+\/(packages\/.+\/.+)/);
  if (match) {
    const [, filePath] = match;
    return filePath;
  }
  return filename;
};
var LEVEL_COLORS = {
  [LogLevel.TRACE]: "gray",
  [LogLevel.DEBUG]: "gray",
  [LogLevel.VERBOSE]: "gray",
  [LogLevel.INFO]: "white",
  [LogLevel.WARN]: "yellow",
  [LogLevel.ERROR]: "red"
};
var truncate = (text, length = 0, right = false) => {
  const str = text && length ? right ? text.slice(-length) : text.substring(0, length) : text ?? "";
  return right ? str.padStart(length, " ") : str.padEnd(length, " ");
};
var DEFAULT_FORMATTER = (config, { path, line, level, message, context, error, scope }) => {
  const column = config.options?.formatter?.column;
  const filepath = path !== void 0 && line !== void 0 ? import_chalk2.default.grey(`${path}:${line}`) : void 0;
  let instance;
  if (scope) {
    const prototype = Object.getPrototypeOf(scope);
    const id = (0, import_util2.getPrototypeSpecificInstanceId)(scope);
    instance = import_chalk2.default.magentaBright(`${prototype.constructor.name}#${id}`);
  }
  const formattedTimestamp = config.options?.formatter?.timestamp ? (/* @__PURE__ */ new Date()).toISOString() : void 0;
  const formattedLevel = import_chalk2.default[LEVEL_COLORS[level]](column ? shortLevelName[level] : LogLevel[level]);
  const padding = column && filepath ? "".padStart(column - filepath.length) : void 0;
  return config.options?.formatter?.timestampFirst ? [
    formattedTimestamp,
    filepath,
    padding,
    formattedLevel,
    instance,
    message,
    context,
    error
  ] : [
    // NOTE: File path must come fist for console hyperlinks.
    // Must not truncate for terminal output.
    filepath,
    padding,
    formattedTimestamp,
    formattedLevel,
    instance,
    message,
    context,
    error
  ];
};
var SHORT_FORMATTER = (config, { path, level, message }) => {
  return [
    import_chalk2.default.grey(truncate(path, 16, true)),
    import_chalk2.default[LEVEL_COLORS[level]](shortLevelName[level]),
    message
  ];
};
var formatter = DEFAULT_FORMATTER;
var CONSOLE_PROCESSOR = (config, entry) => {
  const { level, message, meta, error } = entry;
  if (!shouldLog(entry, config.filters)) {
    return;
  }
  const parts = {
    level,
    message,
    error,
    path: void 0,
    line: void 0,
    scope: void 0,
    context: void 0
  };
  if (meta) {
    parts.path = getRelativeFilename(meta.F);
    parts.line = meta.L;
    parts.scope = meta.S;
  }
  const context = getContextFromEntry(entry);
  if (context) {
    parts.context = (0, import_node_util2.inspect)((0, import_util2.pickBy)(context, (value) => value !== void 0), {
      depth: config.options.depth,
      colors: true,
      maxArrayLength: 8,
      sorted: false
    });
  }
  const line = formatter(config, parts).filter(Boolean).join(" ");
  console.log(line);
};
var DEBUG_PROCESSOR = (config, entry) => {
  console.log((0, import_node_util3.inspect)(entry, false, null, true));
};
var getRelativeFilename2 = (filename) => {
  const match = filename.match(/.+\/(packages\/.+\/.+)/);
  if (match) {
    const [, filePath] = match;
    return filePath;
  }
  return filename;
};
var CONFIG = {
  useTestProcessor: false,
  printFileLinks: false
};
var APP_BROWSER_PROCESSOR = (config, entry) => {
  if (!shouldLog(entry, config.filters)) {
    return;
  }
  const LOG_BROWSER_PREFIX = config.prefix ?? "https://vscode.dev/github.com/dxos/dxos/blob/main/";
  const LOG_BROWSER_CSS = [];
  let link = "";
  if (entry.meta) {
    const filename = getRelativeFilename2(entry.meta.F);
    const filepath = `${LOG_BROWSER_PREFIX.replace(/\/$/, "")}/${filename}`;
    link = `${filepath}#L${entry.meta.L}`;
  }
  let args = [];
  if (entry.meta?.S) {
    const scope = entry.meta?.S;
    const scopeName = scope.name || (0, import_util3.getDebugName)(scope);
    const processPrefix = entry.meta.S?.hostSessionId ? "[worker] " : "";
    args.push(`%c${processPrefix}${scopeName}`, "color:#C026D3;font-weight:bold");
  }
  args.push(entry.message);
  const context = getContextFromEntry(entry);
  if (context) {
    if (Object.keys(context).length === 1 && "error" in context) {
      args.push(context.error);
    } else {
      args.push(context);
    }
  }
  const levels2 = {
    [LogLevel.ERROR]: console.error.bind(console),
    [LogLevel.WARN]: console.warn.bind(console),
    [LogLevel.DEBUG]: console.log.bind(console)
  };
  if (CONFIG.printFileLinks || (0, import_util3.safariCheck)()) {
    if (LOG_BROWSER_CSS?.length) {
      args = [
        `%c${link}
%c${args.join(" ")}`,
        ...LOG_BROWSER_CSS
      ];
    } else {
      args = [
        link + "\n",
        ...args
      ];
    }
  }
  const level = levels2[entry.level] ?? console.log.bind(console);
  if (typeof entry.meta?.C === "function") {
    entry.meta.C(level, args);
  } else {
    level(...args);
  }
};
var TEST_BROWSER_PROCESSOR = (config, entry) => {
  if (!shouldLog(entry, config.filters)) {
    return;
  }
  let path = "";
  if (entry.meta) {
    path = `${getRelativeFilename2(entry.meta.F)}:${entry.meta.L}`;
  }
  let args = [];
  const processPrefix = entry.meta?.S?.hostSessionId ? "[worker] " : "";
  args.push(`${processPrefix}${entry.message}`);
  const context = getContextFromEntry(entry);
  if (context) {
    args.push(context);
  }
  const levels2 = {
    [LogLevel.ERROR]: console.error,
    [LogLevel.WARN]: console.warn,
    [LogLevel.DEBUG]: console.log
  };
  if (CONFIG.printFileLinks) {
    args = [
      path,
      ...args
    ];
  }
  const level = levels2[entry.level] ?? console.log;
  if (typeof entry.meta?.C === "function") {
    entry.meta.C(level, args);
  } else {
    level(...args);
  }
};
var BROWSER_PROCESSOR = CONFIG.useTestProcessor ? TEST_BROWSER_PROCESSOR : APP_BROWSER_PROCESSOR;
var EAGAIN_MAX_DURATION = 1e3;
var createFileProcessor = ({ pathOrFd, levels: levels2, filters }) => {
  let fd;
  return (config, entry) => {
    if (levels2.length > 0 && !levels2.includes(entry.level)) {
      return;
    }
    if (!shouldLog(entry, filters)) {
      return;
    }
    if (typeof pathOrFd === "number") {
      fd = pathOrFd;
    } else {
      try {
        (0, import_node_fs2.mkdirSync)((0, import_node_path.dirname)(pathOrFd));
      } catch {
      }
      fd = (0, import_node_fs2.openSync)(pathOrFd, "a");
    }
    const record = {
      ...entry,
      timestamp: Date.now(),
      ...entry.meta ? {
        meta: {
          file: getRelativeFilename(entry.meta.F),
          line: entry.meta.L
        }
      } : {},
      context: (0, import_util4.jsonlogify)(getContextFromEntry(entry))
    };
    let retryTS = 0;
    while (true) {
      try {
        return (0, import_node_fs2.appendFileSync)(fd, JSON.stringify(record) + "\n");
      } catch (err) {
        if (err.code !== "EAGAIN") {
          throw err;
        }
        if (retryTS === 0) {
          retryTS = performance.now();
        } else {
          if (performance.now() - retryTS > EAGAIN_MAX_DURATION) {
            console.log(`could not write after ${EAGAIN_MAX_DURATION}ms of EAGAIN failures, giving up`);
            throw err;
          }
        }
      }
    }
  };
};
var logFilePath;
var getLogFilePath = () => {
  logFilePath ??= process.env.LOG_FILE ?? (process.env.HOME ? `${process.env.HOME}/.dxlog/${(/* @__PURE__ */ new Date()).toISOString()}.log` : void 0);
  return logFilePath;
};
var FILE_PROCESSOR = createFileProcessor({
  pathOrFd: getLogFilePath(),
  levels: [
    LogLevel.ERROR,
    LogLevel.WARN,
    LogLevel.INFO,
    LogLevel.TRACE
  ]
});
var processors = {
  [LogProcessorType.CONSOLE]: CONSOLE_PROCESSOR,
  [LogProcessorType.BROWSER]: BROWSER_PROCESSOR,
  [LogProcessorType.DEBUG]: DEBUG_PROCESSOR
};
var IS_BROWSER = typeof window !== "undefined" || typeof navigator !== "undefined";
var DEFAULT_PROCESSORS = [
  IS_BROWSER ? BROWSER_PROCESSOR : CONSOLE_PROCESSOR
];
var parseFilter = (filter) => {
  if (typeof filter === "number") {
    return [
      {
        level: filter
      }
    ];
  }
  const parseLogLevel = (level, defValue = LogLevel.WARN) => levels[level.toLowerCase()] ?? defValue;
  const lines = typeof filter === "string" ? filter.split(/,\s*/) : filter;
  return lines.map((filter2) => {
    const [pattern, level] = filter2.split(":");
    return level ? {
      level: parseLogLevel(level),
      pattern
    } : {
      level: parseLogLevel(pattern)
    };
  });
};
var getConfig = (options) => {
  const nodeOptions = "process" in globalThis ? {
    file: process.env.LOG_CONFIG,
    filter: process.env.LOG_FILTER,
    processor: process.env.LOG_PROCESSOR
  } : void 0;
  const mergedOptions = (0, import_lodash2.default)({}, loadOptions(nodeOptions?.file), nodeOptions, options);
  return {
    options: mergedOptions,
    filters: parseFilter(mergedOptions.filter ?? LogLevel.INFO),
    captureFilters: parseFilter(mergedOptions.captureFilter ?? LogLevel.WARN),
    processors: mergedOptions.processor ? [
      processors[mergedOptions.processor]
    ] : DEFAULT_PROCESSORS,
    prefix: mergedOptions.prefix
  };
};
var createLog = () => {
  const log2 = (...params) => processLog(LogLevel.DEBUG, ...params);
  log2._config = getConfig();
  Object.defineProperty(log2, "runtimeConfig", {
    get: () => log2._config
  });
  log2.addProcessor = (processor) => {
    if (DEFAULT_PROCESSORS.filter((p) => p === processor).length === 0) {
      DEFAULT_PROCESSORS.push(processor);
    }
    if (log2._config.processors.filter((p) => p === processor).length === 0) {
      log2._config.processors.push(processor);
    }
  };
  log2.config = (options) => {
    log2._config = getConfig(options);
  };
  log2.trace = (...params) => processLog(LogLevel.TRACE, ...params);
  log2.debug = (...params) => processLog(LogLevel.DEBUG, ...params);
  log2.verbose = (...params) => processLog(LogLevel.VERBOSE, ...params);
  log2.info = (...params) => processLog(LogLevel.INFO, ...params);
  log2.warn = (...params) => processLog(LogLevel.WARN, ...params);
  log2.error = (...params) => processLog(LogLevel.ERROR, ...params);
  log2.catch = (error, context, meta) => processLog(LogLevel.ERROR, error?.message ?? String(error), context, meta, error);
  log2.break = () => log2.info("\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014\u2014");
  log2.stack = (message, context, meta) => processLog(LogLevel.INFO, `${message ?? "Stack Dump"}
${getFormattedStackTrace()}`, context, meta);
  log2.method = createMethodLogDecorator(log2);
  log2.func = createFunctionLogDecorator(log2);
  const processLog = (level, message, context = {}, meta, error) => {
    log2._config.processors.forEach((processor) => processor(log2._config, {
      level,
      message,
      context,
      meta,
      error
    }));
  };
  return log2;
};
var log = globalThis.dx_log ??= createLog();
var start = Date.now();
var last = start;
var debug = (label, args) => {
  const now = Date.now();
  const err = new Error();
  console.group(`DEBUG[${label}]`, JSON.stringify({
    t: Number(now - start).toLocaleString(),
    dt: Number(now - last).toLocaleString(),
    ...args
  }));
  console.warn(err.stack);
  console.groupEnd();
  last = Date.now();
};
var getFormattedStackTrace = () => new Error().stack.split("\n").slice(3).join("\n");
var kOwnershipScope = Symbol("kOwnershipScope");
var kCurrentOwnershipScope = Symbol("kCurrentOwnershipScope");
var kDebugInfoProperties = Symbol("kDebugInfoProperties");
var OwnershipScope = class {
  constructor(constr, parent) {
    this.constr = constr;
    this.parent = parent;
  }
  getInfo() {
    if (!this.instance) {
      return {};
    }
    const props = this.constr.prototype[kDebugInfoProperties] ?? [];
    const info = {};
    for (const prop of props) {
      info[prop] = this.instance[prop];
    }
    return info;
  }
  [import_node_util4.inspect.custom]() {
    return {
      className: this.constr.name,
      info: this.getInfo(),
      parent: this.parent
    };
  }
};
var getCurrentOwnershipScope = (thisRef) => thisRef;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BROWSER_PROCESSOR,
  CONSOLE_PROCESSOR,
  DEBUG_PROCESSOR,
  DEFAULT_FORMATTER,
  FILE_PROCESSOR,
  LogLevel,
  LogProcessorType,
  SHORT_FORMATTER,
  createFileProcessor,
  debug,
  gatherLogInfoFromScope,
  getContextFromEntry,
  getCurrentOwnershipScope,
  getRelativeFilename,
  levels,
  log,
  logInfo,
  omit,
  parseFilter,
  pick,
  shortLevelName,
  shouldLog,
  truncate
});
