// @dxos/protocols/dist/cjs/src/proto/gen/dxos/halo/invitations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthenticationResponse = exports.InvitationOptions = void 0;
var InvitationOptions;
(function (InvitationOptions) {
    /**
     * Defined in:
     *   {@link file://./../../../dxos/halo/invitations.proto}
     */
    let Role;
    (function (Role) {
        Role[Role["GUEST"] = 0] = "GUEST";
        Role[Role["HOST"] = 1] = "HOST";
    })(Role = InvitationOptions.Role || (InvitationOptions.Role = {}));
})(InvitationOptions || (exports.InvitationOptions = InvitationOptions = {}));
var AuthenticationResponse;
(function (AuthenticationResponse) {
    /**
     * Defined in:
     *   {@link file://./../../../dxos/halo/invitations.proto}
     */
    let Status;
    (function (Status) {
        Status[Status["OK"] = 0] = "OK";
        Status[Status["INVALID_OTP"] = 1] = "INVALID_OTP";
        Status[Status["INVALID_OPT_ATTEMPTS"] = 2] = "INVALID_OPT_ATTEMPTS";
        Status[Status["INTERNAL_ERROR"] = 3] = "INTERNAL_ERROR";
        Status[Status["INVALID_SIGNATURE"] = 4] = "INVALID_SIGNATURE";
    })(Status = AuthenticationResponse.Status || (AuthenticationResponse.Status = {}));
})(AuthenticationResponse || (exports.AuthenticationResponse = AuthenticationResponse = {}));
