// @dxos/react-client/dist/lib/node/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var node_exports = {};
__export(node_exports, {
  AgentHostingContext: () => import_chunk_5RUIFUEC.AgentHostingContext,
  AgentHostingProvider: () => import_chunk_5RUIFUEC.AgentHostingProvider,
  ClientContext: () => import_chunk_5RUIFUEC.ClientContext,
  ClientProvider: () => import_chunk_5RUIFUEC.ClientProvider,
  useAgentHostingClient: () => import_chunk_5RUIFUEC.useAgentHostingClient,
  useClient: () => import_chunk_5RUIFUEC.useClient,
  useClientStatus: () => import_chunk_5RUIFUEC.useClientStatus,
  useConfig: () => import_chunk_5RUIFUEC.useConfig,
  useMulticastObservable: () => import_react_hooks.useMulticastObservable,
  useShell: () => import_chunk_5RUIFUEC.useShell,
  useShellDisplay: () => import_chunk_5RUIFUEC.useShellDisplay,
  useStatus: () => import_chunk_5RUIFUEC.useStatus
});
module.exports = __toCommonJS(node_exports);
var import_chunk_5RUIFUEC = require("./chunk-5RUIFUEC.cjs");
__reExport(node_exports, require("@dxos/client"), module.exports);
var import_react_hooks = require("@dxos/react-hooks");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AgentHostingContext,
  AgentHostingProvider,
  ClientContext,
  ClientProvider,
  useAgentHostingClient,
  useClient,
  useClientStatus,
  useConfig,
  useMulticastObservable,
  useShell,
  useShellDisplay,
  useStatus,
  ...require("@dxos/client")
});
